

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>imfusion.util.fusions &mdash; IM-Fusion 0.1+1.g9b7cc62 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="IM-Fusion 0.1+1.g9b7cc62 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> IM-Fusion
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../home.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line_usage.html">Command-line usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../development.html">Development</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../index.html">IM-Fusion</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../index.html">Module code</a> &raquo;</li>
      
    <li>imfusion.util.fusions</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for imfusion.util.fusions</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Module containing functions for annotating gene-transposon fusions.&quot;&quot;&quot;</span>

<span class="c1"># pylint: disable=W0622,W0614,W0401</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">absolute_import</span><span class="p">,</span> <span class="n">division</span><span class="p">,</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">builtins</span> <span class="kn">import</span> <span class="o">*</span>
<span class="c1"># pylint: enable=W0622,W0614,W0401</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">imfusion.model</span> <span class="kn">import</span> <span class="n">Insertion</span>
<span class="kn">from</span> <span class="nn">imfusion.util.tabix</span> <span class="kn">import</span> <span class="n">GtfFile</span>


<div class="viewcode-block" id="annotate_fusions"><a class="viewcode-back" href="../../../api.html#imfusion.util.fusions.annotate_fusions">[docs]</a><span class="k">def</span> <span class="nf">annotate_fusions</span><span class="p">(</span><span class="n">fusions</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">,</span> <span class="n">transposon_features</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Annotates fusions with gene and transposon features.</span>

<span class="sd">    Main function for annotating identified gene-transposon fusions.</span>
<span class="sd">    Adds the following annotations to fusions: gene features, transposon</span>
<span class="sd">    features and the relative orientation of the corresponding insertion</span>
<span class="sd">    with respect to the identified gene.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fusions : List[Fusion]</span>
<span class="sd">        Fusions to annotate.</span>
<span class="sd">    reference_gtf : GtfFile</span>
<span class="sd">        GtfFile instance containing reference gene features.</span>
<span class="sd">    transposon_features : pandas.DataFrame</span>
<span class="sd">        Dataframe containing positions for the features present in</span>
<span class="sd">        the transposon sequence. Used to identify transposon features</span>
<span class="sd">        (such as splice acceptors or donors) that are involved in the</span>
<span class="sd">        identified fusions.</span>

<span class="sd">    Yields</span>
<span class="sd">    -------</span>
<span class="sd">    Fusion</span>
<span class="sd">        Next fusion, annotated with gene/transposon features.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Annotate fusions with features.</span>
    <span class="n">fusions</span> <span class="o">=</span> <span class="n">annotate_with_gene</span><span class="p">(</span><span class="n">fusions</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">)</span>
    <span class="n">fusions</span> <span class="o">=</span> <span class="n">annotate_with_transposon</span><span class="p">(</span><span class="n">fusions</span><span class="p">,</span> <span class="n">transposon_features</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">fusion</span> <span class="ow">in</span> <span class="n">fusions</span><span class="p">:</span>
        <span class="k">yield</span> <span class="n">fusion</span></div>


<span class="k">def</span> <span class="nf">annotate_with_gene</span><span class="p">(</span><span class="n">fusions</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Annotates fusions with genes overlapped by the genomic fusion site.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fusions : List[Fusion]</span>
<span class="sd">        Fusions to annotate.</span>
<span class="sd">    reference_gtf : GtfFile</span>
<span class="sd">        GtfFile instance containing reference gene features.</span>

<span class="sd">    Yields</span>
<span class="sd">    -------</span>
<span class="sd">    Fusion</span>
<span class="sd">        Next fusion, annotated with gene features (includes `gene_id`,</span>
<span class="sd">        `gene_name` and `gene_strand`).</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Get gtf file.</span>
    <span class="c1"># gtf = GtfFile(reference_gtf)</span>

    <span class="c1"># Yield annotated fusions.</span>
    <span class="k">for</span> <span class="n">fusion</span> <span class="ow">in</span> <span class="n">fusions</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">annotated_fusion</span> <span class="ow">in</span> <span class="n">_annotate_genes</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">):</span>
            <span class="k">yield</span> <span class="n">annotated_fusion</span>


<span class="k">def</span> <span class="nf">_annotate_genes</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">ref_gtf</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Annotates a fusion with overlapping genes.&quot;&quot;&quot;</span>

    <span class="c1"># Define region for feature search.</span>
    <span class="k">if</span> <span class="n">fusion</span><span class="o">.</span><span class="n">strand_genome</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">anchor_genome</span> <span class="o">-</span> <span class="n">fusion</span><span class="o">.</span><span class="n">flank_genome</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">anchor_genome</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">anchor_genome</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">anchor_genome</span> <span class="o">+</span> <span class="n">fusion</span><span class="o">.</span><span class="n">flank_genome</span>

    <span class="c1"># Identify overlapped genes.</span>
    <span class="n">exons</span> <span class="o">=</span> <span class="n">ref_gtf</span><span class="o">.</span><span class="n">get_region</span><span class="p">(</span><span class="n">fusion</span><span class="o">.</span><span class="n">seqname</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span>
                               <span class="n">filters</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;feature&#39;</span><span class="p">:</span> <span class="s1">&#39;exon&#39;</span><span class="p">})</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">exons</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Get corresponding gene details.</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">exons</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">],</span> <span class="n">exons</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">],</span> <span class="n">exons</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]))</span>

        <span class="c1"># Return annotated fusions.</span>
        <span class="k">for</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">gene_strand</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">fusion</span><span class="o">.</span><span class="n">_replace</span><span class="p">(</span><span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">gene_name</span><span class="o">=</span><span class="n">gene_name</span><span class="p">,</span>
                                  <span class="n">gene_strand</span><span class="o">=</span><span class="n">_numeric_strand</span><span class="p">(</span><span class="n">gene_strand</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Return original, unannotated fusion.</span>
        <span class="k">yield</span> <span class="n">fusion</span>



<span class="k">def</span> <span class="nf">_numeric_strand</span><span class="p">(</span><span class="n">strand</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">1</span>
    <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="o">-</span><span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>


<span class="k">def</span> <span class="nf">annotate_with_transposon</span><span class="p">(</span><span class="n">fusions</span><span class="p">,</span> <span class="n">transposon_features</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Annotates fusions with transposon features overlapped by the fusion.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fusions : List[Fusion]</span>
<span class="sd">        Fusions to annotate, in DataFrame format.</span>
<span class="sd">    transposon_features : pandas.DataFrame</span>
<span class="sd">        Dataframe containing positions for the features present in</span>
<span class="sd">        the transposon sequence. Used to identify transposon features</span>
<span class="sd">        (such as splice acceptors or donors) that are involved in the</span>
<span class="sd">        identified fusions.</span>

<span class="sd">    Yields</span>
<span class="sd">    -------</span>
<span class="sd">    Fusion</span>
<span class="sd">        Next fusion, annotated with transposon features.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">fusion</span> <span class="ow">in</span> <span class="n">fusions</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">annotated_fusion</span> <span class="ow">in</span> <span class="n">_annotate_features</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">transposon_features</span><span class="p">):</span>
            <span class="k">yield</span> <span class="n">annotated_fusion</span>


<span class="k">def</span> <span class="nf">_annotate_features</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">transposon_features</span><span class="p">):</span>
    <span class="c1"># Identify overlapped features.</span>
    <span class="n">features</span> <span class="o">=</span> <span class="n">transposon_features</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span>
        <span class="p">(</span><span class="n">transposon_features</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">fusion</span><span class="o">.</span><span class="n">anchor_transposon</span><span class="p">)</span> <span class="o">&amp;</span>
        <span class="p">(</span><span class="n">transposon_features</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">fusion</span><span class="o">.</span><span class="n">anchor_transposon</span><span class="p">)]</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">features</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">features</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="k">yield</span> <span class="n">fusion</span><span class="o">.</span><span class="n">_replace</span><span class="p">(</span><span class="n">feature_name</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">],</span>
                                  <span class="n">feature_type</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">],</span>
                                  <span class="n">feature_strand</span><span class="o">=</span><span class="n">feature</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">yield</span> <span class="n">fusion</span>


<div class="viewcode-block" id="place_fusions"><a class="viewcode-back" href="../../../api.html#imfusion.util.fusions.place_fusions">[docs]</a><span class="k">def</span> <span class="nf">place_fusions</span><span class="p">(</span><span class="n">fusions</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">max_dist</span><span class="o">=</span><span class="mi">5000</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Derives insertions by placing fusions at approximate genomic locations.</span>

<span class="sd">    Main function for deriving insertions from annotated gene-transposon</span>
<span class="sd">    fusions. Derives insertions by determining an approximate genomic</span>
<span class="sd">    location that is compatible with the gene/transposon feature annotations</span>
<span class="sd">    of the fusions. Fusions are therefore expected to be properly annotated</span>
<span class="sd">    for gene/transposon features.</span>

<span class="sd">    An insertion is essentially &#39;placed&#39; by looking for the first genomic</span>
<span class="sd">    position that does not overlap with a reference feature, in the direction</span>
<span class="sd">    that is compatible with the insertions orientation w.r.t. its target gene.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fusions : List[Fusion]</span>
<span class="sd">        List of fusions to convert.</span>
<span class="sd">    sample_id : str</span>
<span class="sd">        Sample id that should be used for the insertions.</span>
<span class="sd">    reference_gtf : GtfFile</span>
<span class="sd">        GtfFile containing the reference features. Expected to conform</span>
<span class="sd">        to the Ensembl reference gtf format.</span>
<span class="sd">    offset : int</span>
<span class="sd">        Minimum offset of the transposon to the closest reference gene feature.</span>
<span class="sd">    max_dist : int</span>
<span class="sd">        Maximum distance that an insertion may be placed from the</span>
<span class="sd">        genomic anchor of the fusion.</span>

<span class="sd">    Yields</span>
<span class="sd">    -------</span>
<span class="sd">    Insertion</span>
<span class="sd">        Next insertion derived from the given fusions.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">fusion</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fusions</span><span class="p">):</span>
        <span class="k">yield</span> <span class="n">place_fusion</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">,</span> <span class="n">offset</span><span class="p">,</span>
                           <span class="n">max_dist</span><span class="p">,</span> <span class="n">id_</span><span class="o">=</span><span class="s1">&#39;INS_{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span></div>


<span class="k">def</span> <span class="nf">place_fusion</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">,</span>
                 <span class="n">offset</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">max_dist</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">id_</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Derives an insertion from a given fusion.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fusion : Fusion</span>
<span class="sd">        Fusions to derive an insertion for.</span>
<span class="sd">    sample_id : str</span>
<span class="sd">        Sample id that should be used for the insertion.</span>
<span class="sd">    reference_gtf : GtfFile</span>
<span class="sd">        GtfFile containing the reference features. Expected to conform</span>
<span class="sd">        to the Ensembl reference gtf format.</span>
<span class="sd">    offset : int</span>
<span class="sd">        Minimum offset of the transposon to the closest reference gene feature.</span>
<span class="sd">    max_dist : int</span>
<span class="sd">        Maximum distance that an insertion may be placed from the</span>
<span class="sd">        genomic anchor of the fusion.</span>
<span class="sd">    id_: str</span>
<span class="sd">        ID to use for the insertion.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Insertion</span>
<span class="sd">        The insertion derived from the given fusion.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">strand</span><span class="p">,</span> <span class="n">orientation</span> <span class="o">=</span> <span class="n">insertion_orientation</span><span class="p">(</span><span class="n">fusion</span><span class="p">)</span>
    <span class="n">position</span> <span class="o">=</span> <span class="n">insertion_position</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">orientation</span><span class="p">,</span>
                                  <span class="n">offset</span><span class="o">=</span><span class="n">offset</span><span class="p">,</span> <span class="n">max_dist</span><span class="o">=</span><span class="n">max_dist</span><span class="p">)</span>

    <span class="n">insertion</span> <span class="o">=</span> <span class="n">Insertion</span><span class="p">(</span>
        <span class="nb">id</span><span class="o">=</span><span class="n">id_</span><span class="p">,</span>
        <span class="n">seqname</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">seqname</span><span class="p">,</span>
        <span class="n">position</span><span class="o">=</span><span class="n">position</span><span class="p">,</span>
        <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span>
        <span class="n">sample_id</span><span class="o">=</span><span class="n">sample_id</span><span class="p">,</span>
        <span class="n">gene_id</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">gene_id</span><span class="p">,</span>
        <span class="n">gene_name</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">gene_name</span><span class="p">,</span>
        <span class="n">gene_strand</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">gene_strand</span><span class="p">,</span>
        <span class="n">orientation</span><span class="o">=</span><span class="n">orientation</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">feature_type</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">feature_type</span><span class="p">,</span>
        <span class="n">feature_strand</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">feature_strand</span><span class="p">,</span>
        <span class="n">anchor_genome</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">anchor_genome</span><span class="p">,</span>
        <span class="n">anchor_transposon</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">anchor_transposon</span><span class="p">,</span>
        <span class="n">flank_genome</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">flank_genome</span><span class="p">,</span>
        <span class="n">flank_transposon</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">flank_transposon</span><span class="p">,</span>
        <span class="n">spanning_reads</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">spanning_reads</span><span class="p">,</span>
        <span class="n">supporting_mates</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">supporting_mates</span><span class="p">,</span>
        <span class="n">spanning_mates</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">spanning_mates</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">insertion</span>


<span class="k">def</span> <span class="nf">insertion_position</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">orientation</span><span class="p">,</span>
                       <span class="n">offset</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">max_dist</span><span class="o">=</span><span class="mi">5000</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Determines an approximate genomic location for an RNA-seq insertion.</span>

<span class="sd">    This position is determined by looking for the first genomic</span>
<span class="sd">    position that does not overlap with a reference feature, in the direction</span>
<span class="sd">    that is compatible with the insertions orientation w.r.t. its target gene.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fusions : Fusion</span>
<span class="sd">        Fusion to find a position fore.</span>
<span class="sd">    reference_gtf : GtfFile</span>
<span class="sd">        GtfFile containing the reference features. Expected to conform</span>
<span class="sd">        to the Ensembl reference gtf format.</span>
<span class="sd">    strand : int</span>
<span class="sd">        Strand of the corresponding insertion (1 or -1).</span>
<span class="sd">    orientation : str</span>
<span class="sd">        Orientation of the corresponding insertion with respect to the</span>
<span class="sd">        annotated gene (sense or antisense).</span>
<span class="sd">    offset : int</span>
<span class="sd">        Minimum offset of the transposon to the closest reference gene feature.</span>
<span class="sd">    max_dist : int</span>
<span class="sd">        Maximum distance that an insertion may be placed from the</span>
<span class="sd">        genomic anchor of the fusion.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        Approximate position for the insertion.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Determine SD/SA type.</span>
    <span class="n">splice_type</span> <span class="o">=</span> <span class="n">_get_splice_type</span><span class="p">(</span><span class="n">fusion</span><span class="p">)</span>

    <span class="c1"># Determine direction in which to place insertion. Here the orientation</span>
    <span class="c1"># and insertion strand is used to first determine the orientation of</span>
    <span class="c1"># the gene, and the splice type then determines the actual direction.</span>
    <span class="n">direction</span> <span class="o">=</span> <span class="p">((</span><span class="o">-</span><span class="mi">1</span> <span class="k">if</span> <span class="n">splice_type</span> <span class="o">==</span> <span class="s1">&#39;SD&#39;</span> <span class="k">else</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span>
                 <span class="p">(</span><span class="mi">1</span> <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;sense&#39;</span> <span class="k">else</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">strand</span><span class="p">)</span>

    <span class="c1"># Search for a suitable position, starting from the initial anchor.</span>
    <span class="n">anchor_location</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">anchor_genome</span>
    <span class="n">location</span> <span class="o">=</span> <span class="n">anchor_location</span> <span class="o">+</span> <span class="p">(</span><span class="n">direction</span> <span class="o">*</span> <span class="n">offset</span><span class="p">)</span>

    <span class="c1"># Perform the search, within a specified maximum distance.</span>
    <span class="n">converged</span> <span class="o">=</span> <span class="bp">False</span>

    <span class="k">while</span> <span class="ow">not</span> <span class="n">converged</span> <span class="ow">and</span> <span class="nb">abs</span><span class="p">(</span><span class="n">location</span> <span class="o">-</span> <span class="n">anchor_location</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">max_dist</span><span class="p">:</span>
        <span class="c1"># Check if no features are within &#39;offset&#39; distance.</span>
        <span class="n">bounds</span> <span class="o">=</span> <span class="n">location</span><span class="p">,</span> <span class="n">location</span> <span class="o">+</span> <span class="p">((</span><span class="n">direction</span> <span class="o">*</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">offset</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">bounds</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">bounds</span><span class="p">)</span>

        <span class="n">overlap</span> <span class="o">=</span> <span class="n">reference_gtf</span><span class="o">.</span><span class="n">get_region</span><span class="p">(</span><span class="n">fusion</span><span class="o">.</span><span class="n">seqname</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span>
                                           <span class="n">filters</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">feature</span><span class="o">=</span><span class="s1">&#39;exon&#39;</span><span class="p">))</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">overlap</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Advance our position to avoid the encountered overlap.</span>
            <span class="k">if</span> <span class="n">direction</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># Move in forward direction.</span>
                <span class="n">location</span> <span class="o">=</span> <span class="n">overlap</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="n">offset</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Move in reverse direction.</span>
                <span class="n">location</span> <span class="o">=</span> <span class="n">overlap</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="n">offset</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># We have found our position, so stop search.</span>
            <span class="n">converged</span> <span class="o">=</span> <span class="bp">True</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">converged</span><span class="p">:</span>
        <span class="n">location</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

    <span class="k">return</span> <span class="n">location</span>


<span class="k">def</span> <span class="nf">_get_splice_type</span><span class="p">(</span><span class="n">fusion</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Determines the splice type (SA/SD) of a fusion.&quot;&quot;&quot;</span>

    <span class="n">feature</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">feature_name</span>

    <span class="k">if</span> <span class="n">feature</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Fusion does not have an annotated transposon feature&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">feature</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;SA&#39;</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;SA&#39;</span>
    <span class="k">elif</span> <span class="n">feature</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;SD&#39;</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;SD&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Unknown splice type for feature {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">feature</span><span class="p">))</span>


<span class="k">def</span> <span class="nf">insertion_orientation</span><span class="p">(</span><span class="n">fusion</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Identifies the strand and orientation of the corresponding insertion.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fusion : Fusion</span>
<span class="sd">        Fusion to identify the strand and orientation for.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    tuple(int, str)</span>
<span class="sd">        Strand and orientation of the insertion corresponding</span>
<span class="sd">        with the given fusion.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">strand</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">strand_genome</span> <span class="o">*</span> <span class="n">fusion</span><span class="o">.</span><span class="n">strand_transposon</span>

    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">fusion</span><span class="o">.</span><span class="n">gene_strand</span><span class="p">):</span>
        <span class="n">orientation</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="n">fusion</span><span class="o">.</span><span class="n">gene_strand</span><span class="p">:</span>
        <span class="n">orientation</span> <span class="o">=</span> <span class="s1">&#39;sense&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">orientation</span> <span class="o">=</span> <span class="s1">&#39;antisense&#39;</span>

    <span class="k">return</span> <span class="n">strand</span><span class="p">,</span> <span class="n">orientation</span>
</pre></div>

           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Julian de Ruiter.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.1+1.g9b7cc62',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>