

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>imfusion.util.reference &mdash; IM-Fusion 0.1+1.g9b7cc62 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="IM-Fusion 0.1+1.g9b7cc62 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> IM-Fusion
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../home.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../command_line_usage.html">Command-line usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../development.html">Development</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../index.html">IM-Fusion</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../index.html">Module code</a> &raquo;</li>
      
    <li>imfusion.util.reference</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for imfusion.util.reference</h1><div class="highlight"><pre>
<span></span><span class="c1"># pylint: disable=W0622,W0614,W0401</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">absolute_import</span><span class="p">,</span> <span class="n">division</span><span class="p">,</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">builtins</span> <span class="kn">import</span> <span class="o">*</span>
<span class="c1"># pylint: enable=W0622,W0614,W0401</span>

<span class="kn">import</span> <span class="nn">pyfaidx</span>

<span class="kn">from</span> <span class="nn">imfusion.util.tabix</span> <span class="kn">import</span> <span class="n">GtfFile</span>


<div class="viewcode-block" id="concatenate_fastas"><a class="viewcode-back" href="../../../api.html#imfusion.util.reference.concatenate_fastas">[docs]</a><span class="k">def</span> <span class="nf">concatenate_fastas</span><span class="p">(</span><span class="n">fasta_paths</span><span class="p">,</span> <span class="n">output_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Concatenates multiple fasta files into a single file.</span>

<span class="sd">    This function combines multiple fasta files into a single output</span>
<span class="sd">    file. It is mainly used to generate the combined reference genome</span>
<span class="sd">    that contains both the reference genome sequence and the</span>
<span class="sd">    transposon sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fasta_paths : List[pathlib.Path]</span>
<span class="sd">        Paths to fasta files that should be concatentated.</span>
<span class="sd">    output_path : pathlib.Paths</span>
<span class="sd">        Path for the combined output file.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="n">output_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">fasta_path</span> <span class="ow">in</span> <span class="n">fasta_paths</span><span class="p">:</span>
            <span class="c1"># Open fasta file.</span>
            <span class="n">fasta</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Fasta</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">))</span>

            <span class="c1"># Read/write sequences.</span>
            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta</span><span class="p">:</span>
                <span class="n">_write_fasta_record</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">long_name</span><span class="p">,</span>
                                    <span class="n">sequence</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="p">))</span></div>


<span class="k">def</span> <span class="nf">_write_fasta_record</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">max_width</span><span class="o">=</span><span class="mi">80</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Writes a fastq record to an opened file.&quot;&quot;&quot;</span>

    <span class="c1"># Group sequence into chunks of max_width.</span>
    <span class="n">chunked</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="n">sequence</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">max_width</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span>
                        <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="n">max_width</span><span class="p">)))</span>

    <span class="c1"># Write record to file.</span>
    <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">u&#39;&gt;{}</span><span class="se">\n</span><span class="s1">{}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">chunked</span><span class="p">))</span>


<div class="viewcode-block" id="mask_regions"><a class="viewcode-back" href="../../../api.html#imfusion.util.reference.mask_regions">[docs]</a><span class="k">def</span> <span class="nf">mask_regions</span><span class="p">(</span><span class="n">refseq_path</span><span class="p">,</span> <span class="n">blacklist_regions</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Masks blacklisted regions in a given reference sequence.</span>

<span class="sd">    This function removes blacklisted regions from a given reference genome.</span>
<span class="sd">    Blacklisted regions are removed by replacing their original sequence</span>
<span class="sd">    with a sequences of &#39;N&#39; nucleotides.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    refseq_path : pathlib.Path</span>
<span class="sd">        Path to the reference sequence in fasta format.</span>
<span class="sd">    blacklist_regions : List[tuple(str, int, int)]:</span>
<span class="sd">        List of regions that should be blacklisted in the reference</span>
<span class="sd">        sequence. Regions should be specified as a tuple of (chromosome,</span>
<span class="sd">        start_position, end_position). For example: (&#39;1&#39;, 2000, 2200).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pathlib.Path</span>
<span class="sd">        Path to edited reference sequence.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Open the reference as a mutable sequence.</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Fasta</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">refseq_path</span><span class="p">),</span> <span class="n">mutable</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># Mask regions.</span>
    <span class="k">for</span> <span class="n">seqname</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="n">blacklist_regions</span><span class="p">:</span>
        <span class="n">ref</span><span class="p">[</span><span class="n">seqname</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;N&#39;</span> <span class="o">*</span> <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span>

    <span class="c1"># Close file.</span>
    <span class="n">ref</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">refseq_path</span></div>


<div class="viewcode-block" id="blacklist_for_regions"><a class="viewcode-back" href="../../../api.html#imfusion.util.reference.blacklist_for_regions">[docs]</a><span class="k">def</span> <span class="nf">blacklist_for_regions</span><span class="p">(</span><span class="n">region_strs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Builds blacklist region list for region strings.</span>

<span class="sd">    Parses region strings into a list of blacklist region tuples.</span>
<span class="sd">    Region strings should be provided in the following format:</span>
<span class="sd">    &#39;chromosome:start-end&#39;. For example, &#39;X:1000-2000&#39; denotes a</span>
<span class="sd">    a region on chromosome X from position 1000 to 2000.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    region_strs : List[str]</span>
<span class="sd">        List of region strings.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    tuple(str, int, int)</span>
<span class="sd">        List of blacklist region tuples.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="p">[</span><span class="n">_parse_region_str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">region_strs</span><span class="p">]</span></div>


<span class="k">def</span> <span class="nf">_parse_region_str</span><span class="p">(</span><span class="n">region_str</span><span class="p">):</span>
    <span class="n">seqname</span><span class="p">,</span> <span class="n">range_str</span> <span class="o">=</span> <span class="n">region_str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)</span>
    <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">e</span><span class="p">)</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">range_str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">seqname</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span>


<div class="viewcode-block" id="blacklist_for_genes"><a class="viewcode-back" href="../../../api.html#imfusion.util.reference.blacklist_for_genes">[docs]</a><span class="k">def</span> <span class="nf">blacklist_for_genes</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Builds blacklist frame for given genes.</span>

<span class="sd">    Returns a list of blacklist regions encompassing the regions</span>
<span class="sd">    spanned by the genes corresponding to the given gene ids.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene_ids : List[str]</span>
<span class="sd">        List of (Ensembl) gene ids.</span>
<span class="sd">    reference_gtf : GtfFile</span>
<span class="sd">        GtfFile instance containing reference gene features.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    tuple(str, int, int)</span>
<span class="sd">        List of blacklist region tuples.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="p">[</span><span class="n">_get_gene_region</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">]</span></div>


<span class="k">def</span> <span class="nf">_get_gene_region</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">):</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="n">reference_gtf</span><span class="o">.</span><span class="n">get_gene</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Julian de Ruiter.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.1+1.g9b7cc62',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>