

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>API &mdash; IM-Fusion 0.3.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="IM-Fusion 0.3.0 documentation" href="index.html"/>
        <link rel="next" title="Extras" href="extras.html"/>
        <link rel="prev" title="Usage" href="usage.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> IM-Fusion
          

          
          </a>

          
            
            
              <div class="version">
                0.3.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">IM-Fusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#building-references">Building references</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#indexers">Indexers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#identifying-insertions">Identifying insertions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#differential-expression">Differential expression</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#quantification">Quantification</a></li>
<li class="toctree-l3"><a class="reference internal" href="#de-tests">DE tests</a></li>
<li class="toctree-l3"><a class="reference internal" href="#de-results">DE results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extras.html">Extras</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">IM-Fusion</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>API</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="api">
<h1>API<a class="headerlink" href="#api" title="Permalink to this headline">¶</a></h1>
<div class="section" id="building-references">
<h2>Building references<a class="headerlink" href="#building-references" title="Permalink to this headline">¶</a></h2>
<div class="section" id="indexers">
<h3>Indexers<a class="headerlink" href="#indexers" title="Permalink to this headline">¶</a></h3>
<p>IM-Fusion uses <code class="docutils literal"><span class="pre">Indexer</span></code> classes to generate references for the
different RNA-seq aligners, with one class provided for each aligner. The main
methods provided by the Indexers are the <code class="docutils literal"><span class="pre">check_dependencies</span></code> and <code class="docutils literal"><span class="pre">build</span></code>
methods. The former is used to check whether required external dependencies
are present in <code class="docutils literal"><span class="pre">$PATH</span></code>, whilst the latter should be called with the required
reference files to build the actual reference.</p>
<dl class="class">
<dt id="imfusion.build.indexers.Indexer">
<em class="property">class </em><code class="descclassname">imfusion.build.indexers.</code><code class="descname">Indexer</code><span class="sig-paren">(</span><em>logger=None</em>, <em>skip_index=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/base.html#Indexer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.Indexer" title="Permalink to this definition">¶</a></dt>
<dd><p>Base Indexer class.</p>
<p>This base class defines the interface for Indexer classes and provides
basic functionality for building augmented reference genomes. Functionality
for building the indices for alignment is provided by sub-classes
(as this is aligner specific), typically by overriding the protected
<code class="docutils literal"><span class="pre">_build_indices</span></code> method.</p>
<p>The main interface of the class is formed by the <code class="docutils literal"><span class="pre">check_dependencies</span></code>
and <code class="docutils literal"><span class="pre">build</span></code> methods. The former is used to check whether required
external dependencies are present in <code class="docutils literal"><span class="pre">$PATH</span></code>, whilst the latter is
called with the required reference files to build the actual reference.
On completion, the <code class="docutils literal"><span class="pre">build</span></code> method returns a <code class="docutils literal"><span class="pre">Reference</span></code> object, which
defines paths to the various reference files. The class of this object is
defined by the <code class="docutils literal"><span class="pre">_reference_class</span></code> method, which may be overridden to
return aligner-specific <code class="docutils literal"><span class="pre">Reference</span></code> subclasses.</p>
<p>Additionally, the class provides the methods <code class="docutils literal"><span class="pre">configure_args</span></code> and
<code class="docutils literal"><span class="pre">from_args</span></code>, which are used to configure the argument parser for
<code class="docutils literal"><span class="pre">imfusion-build</span></code> and to instantate instances from parsed arguments.
Subclasses with extra options should override the protected <code class="docutils literal"><span class="pre">_parse_args</span></code>
method, which extracts constructor parameters from argparse arguments
for <code class="docutils literal"><span class="pre">from_args</span></code>, allowing for flexible interpretation of these arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>logger</strong> (<em>logging.Logger</em>) &#8211; Logger to be used for logging messages.</li>
<li><strong>skip_index</strong> (<em>bool</em>) &#8211; Whether to skip building the index. Mostly used for debugging purposes,
as this typically results in an incomplete reference.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="imfusion.build.indexers.Indexer.build">
<code class="descname">build</code><span class="sig-paren">(</span><em>refseq_path</em>, <em>gtf_path</em>, <em>transposon_path</em>, <em>transposon_features_path</em>, <em>output_dir</em>, <em>blacklist_regions=None</em>, <em>blacklist_genes=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/base.html#Indexer.build"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.Indexer.build" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds an augmented reference containing the transposon sequence.</p>
<p>The reference is built by first copying the original reference
sequence into the output directory and then augmenting this reference
with the transposon sequence. If needed, blacklisted genes/regions are
then masked within this reference. Next, other supporting files
(such as the reference GTF) are copied and indexed as needed. As a
final step, aligner-specific indices are built for the augmented
reference.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>refseq_path</strong> (<em>Path</em>) &#8211; Path to the reference sequence (in Fasta format).</li>
<li><strong>gtf_path</strong> (<em>Path</em>) &#8211; Path to the reference GTF file.</li>
<li><strong>transposon_path</strong> (<em>Path</em>) &#8211; Path to the transposon sequence (in Fasta format).</li>
<li><strong>transposon_features_path</strong> (<em>Path</em>) &#8211; Path to the transposon features (tab-separated file).</li>
<li><strong>output_dir</strong> (<em>Path</em>) &#8211; Path to the output directory. Should not yet exist.</li>
<li><strong>blacklist_regions</strong> (<em>List</em><em>[</em><em>Tuple</em><em>[</em><em>str</em><em>, </em><em>int</em><em>, </em><em>int</em><em>]</em><em></em><em>]</em><em></em>) &#8211; List or regions that should be blacklisted, specified as a tuple
of (chromosome, start, end).</li>
<li><strong>blacklist_genes</strong> (<em>List</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; List of genes that should be blacklisted. Should correspond to
<code class="docutils literal"><span class="pre">gene_id</span></code> entries in the passed GTF file.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Returns a <code class="docutils literal"><span class="pre">Reference</span></code> object, which describes the paths to
various files within the generated reference.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#imfusion.build.indexers.Reference" title="imfusion.build.indexers.Reference">Reference</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="imfusion.build.indexers.Indexer.check_dependencies">
<code class="descname">check_dependencies</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/base.html#Indexer.check_dependencies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.Indexer.check_dependencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if all required external dependencies are in <code class="docutils literal"><span class="pre">$PATH</span></code>.</p>
<p>Raises a ValueError if any dependencies are missing.</p>
</dd></dl>

<dl class="classmethod">
<dt id="imfusion.build.indexers.Indexer.configure_args">
<em class="property">classmethod </em><code class="descname">configure_args</code><span class="sig-paren">(</span><em>parser</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/base.html#Indexer.configure_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.Indexer.configure_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Configures an argument parser for the Indexer.</p>
<p>Used by <code class="docutils literal"><span class="pre">imfusion-build</span></code> to configure the sub-command for
this indexer (if registered as an Indexer using the
<code class="docutils literal"><span class="pre">register_indexer</span></code> function).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>parser</strong> (<em>argparse.ArgumentParser</em>) &#8211; Argument parser to configure.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.Indexer.dependencies">
<code class="descname">dependencies</code><a class="headerlink" href="#imfusion.build.indexers.Indexer.dependencies" title="Permalink to this definition">¶</a></dt>
<dd><p>External dependencies required by Indexer.</p>
</dd></dl>

<dl class="classmethod">
<dt id="imfusion.build.indexers.Indexer.from_args">
<em class="property">classmethod </em><code class="descname">from_args</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/base.html#Indexer.from_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.Indexer.from_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs an Indexer instance from given arguments.</p>
<p>Instantiates an Indexer instance from the given argparse arguments.
Uses the <code class="docutils literal"><span class="pre">_parse_args</span></code> method internally, which performs the actual
argument-to-parameter extraction. As such, the <code class="docutils literal"><span class="pre">_parse_args</span></code> method
should be overridden in any subclasses with extra parameters.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<em>argparse.Namespace</em>) &#8211; Arguments to parse.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="imfusion.build.indexers.StarIndexer">
<em class="property">class </em><code class="descclassname">imfusion.build.indexers.</code><code class="descname">StarIndexer</code><span class="sig-paren">(</span><em>logger=None</em>, <em>skip_index=False</em>, <em>overhang=100</em>, <em>threads=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/star.html#StarIndexer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.StarIndexer" title="Permalink to this definition">¶</a></dt>
<dd><p>Indexer that builds references for the STAR aligner.</p>
<p>Performs the same steps as the base <code class="docutils literal"><span class="pre">Indexer</span></code> class, but additionally
generates an index for alignment with STAR using STAR&#8217;s <code class="docutils literal"><span class="pre">genomeGenerate</span></code>
command. Special attention should be paid to the <code class="docutils literal"><span class="pre">overhang</span></code> parameter,
which defines the overhang used by STAR in the build reference
(see the <code class="docutils literal"><span class="pre">sjdbOverhang</span></code> parameter in the STAR documentation for more
details). Ideally, the value for this parameter should be one less than the
length of the used reads.</p>
<dl class="classmethod">
<dt id="imfusion.build.indexers.StarIndexer.configure_args">
<em class="property">classmethod </em><code class="descname">configure_args</code><span class="sig-paren">(</span><em>parser</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/star.html#StarIndexer.configure_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.StarIndexer.configure_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Configures an argument parser for the Indexer.</p>
<p>Used by <code class="docutils literal"><span class="pre">imfusion-build</span></code> to configure the sub-command for
this indexer (if registered as an Indexer using the
<code class="docutils literal"><span class="pre">register_indexer</span></code> function).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>parser</strong> (<em>argparse.ArgumentParser</em>) &#8211; Argument parser to configure.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.StarIndexer.dependencies">
<code class="descname">dependencies</code><a class="headerlink" href="#imfusion.build.indexers.StarIndexer.dependencies" title="Permalink to this definition">¶</a></dt>
<dd><p>External dependencies required by this indexer.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="imfusion.build.indexers.TophatIndexer">
<em class="property">class </em><code class="descclassname">imfusion.build.indexers.</code><code class="descname">TophatIndexer</code><span class="sig-paren">(</span><em>logger=None</em>, <em>skip_index=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/tophat.html#TophatIndexer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.TophatIndexer" title="Permalink to this definition">¶</a></dt>
<dd><p>Indexer that builds references for the Tophat-Fusion aligner.</p>
<p>Performs the same steps as the base <code class="docutils literal"><span class="pre">Indexer</span></code> class, but additionally
generates an index for alignment with Tophat-Fusion using Bowtie and
Tophat2.</p>
<dl class="attribute">
<dt id="imfusion.build.indexers.TophatIndexer.dependencies">
<code class="descname">dependencies</code><a class="headerlink" href="#imfusion.build.indexers.TophatIndexer.dependencies" title="Permalink to this definition">¶</a></dt>
<dd><p>External dependencies required by this indexer.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">Indexer</span></code> classes return <code class="docutils literal"><span class="pre">Reference</span></code> objects after building
references, which define paths to specific files within the reference.
This includes, for example, paths to the augmented fasta file, to the file
describing transposon features and to the reference index. The <code class="docutils literal"><span class="pre">Reference</span></code>
base class defines paths that should be available for all references.
Subclasses may define additional paths that are specific to the
corresponding aligner.</p>
<dl class="class">
<dt id="imfusion.build.indexers.Reference">
<em class="property">class </em><code class="descclassname">imfusion.build.indexers.</code><code class="descname">Reference</code><span class="sig-paren">(</span><em>reference_path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/base.html#Reference"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.Reference" title="Permalink to this definition">¶</a></dt>
<dd><p>Base Reference class.</p>
<p>Reference classes define paths to specific files within the reference. This
includes, for example, paths to the augmented fasta file and to the r
eference index. This base class defines paths that should be available
for all references. Subclasses may define additional paths that are
specific to the corresponding aligner.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>reference_path</strong> (<em>Path</em>) &#8211; Path to the reference directory.</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="imfusion.build.indexers.Reference.base_path">
<code class="descname">base_path</code><a class="headerlink" href="#imfusion.build.indexers.Reference.base_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to reference base directory.</p>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.Reference.exon_gtf_path">
<code class="descname">exon_gtf_path</code><a class="headerlink" href="#imfusion.build.indexers.Reference.exon_gtf_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to exon gtf.</p>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.Reference.fasta_path">
<code class="descname">fasta_path</code><a class="headerlink" href="#imfusion.build.indexers.Reference.fasta_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to reference sequence.</p>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.Reference.features_path">
<code class="descname">features_path</code><a class="headerlink" href="#imfusion.build.indexers.Reference.features_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to transposon features.</p>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.Reference.gtf_path">
<code class="descname">gtf_path</code><a class="headerlink" href="#imfusion.build.indexers.Reference.gtf_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to reference gtf.</p>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.Reference.index_path">
<code class="descname">index_path</code><a class="headerlink" href="#imfusion.build.indexers.Reference.index_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to index.</p>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.Reference.indexed_gtf_path">
<code class="descname">indexed_gtf_path</code><a class="headerlink" href="#imfusion.build.indexers.Reference.indexed_gtf_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to reference gtf.</p>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.Reference.transposon_name">
<code class="descname">transposon_name</code><a class="headerlink" href="#imfusion.build.indexers.Reference.transposon_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of transposon sequence.</p>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.Reference.transposon_path">
<code class="descname">transposon_path</code><a class="headerlink" href="#imfusion.build.indexers.Reference.transposon_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of transposon sequence.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="imfusion.build.indexers.StarReference">
<em class="property">class </em><code class="descclassname">imfusion.build.indexers.</code><code class="descname">StarReference</code><span class="sig-paren">(</span><em>reference_path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/star.html#StarReference"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.StarReference" title="Permalink to this definition">¶</a></dt>
<dd><p>Star Reference class.</p>
<p>Defines paths to files within the STAR reference. Currently the same
as the base <code class="docutils literal"><span class="pre">Reference</span></code> class.</p>
</dd></dl>

<dl class="class">
<dt id="imfusion.build.indexers.TophatReference">
<em class="property">class </em><code class="descclassname">imfusion.build.indexers.</code><code class="descname">TophatReference</code><span class="sig-paren">(</span><em>reference_path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/build/indexers/tophat.html#TophatReference"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.build.indexers.TophatReference" title="Permalink to this definition">¶</a></dt>
<dd><p>Tophat Reference class.</p>
<p>Defines paths to files within the Tophat-Fusion reference. Compared to the
base reference, this class adds an additional path to the transcriptome
index that is used by Tophat during alignment.</p>
<dl class="attribute">
<dt id="imfusion.build.indexers.TophatReference.index_path">
<code class="descname">index_path</code><a class="headerlink" href="#imfusion.build.indexers.TophatReference.index_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to index.</p>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.build.indexers.TophatReference.transcriptome_path">
<code class="descname">transcriptome_path</code><a class="headerlink" href="#imfusion.build.indexers.TophatReference.transcriptome_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to transcriptome index.</p>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="identifying-insertions">
<h2>Identifying insertions<a class="headerlink" href="#identifying-insertions" title="Permalink to this headline">¶</a></h2>
<p>IM-Fusion uses <code class="docutils literal"><span class="pre">Aligner</span></code> classes to identify insertions using the different
RNA-seq aligners, with one class provided for each supported aligner. The main
methods provided by the Indexers are the <code class="docutils literal"><span class="pre">check_dependencies</span></code> and
<code class="docutils literal"><span class="pre">identify_insertions</span></code> methods. The latter should be called with a reference
instance and the sequence read files to identify insertions.</p>
<dl class="class">
<dt id="imfusion.insertions.aligners.Aligner">
<em class="property">class </em><code class="descclassname">imfusion.insertions.aligners.</code><code class="descname">Aligner</code><span class="sig-paren">(</span><em>reference</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/base.html#Aligner"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.Aligner" title="Permalink to this definition">¶</a></dt>
<dd><p>Base aligner class.</p>
<p>This base class defines the interface for <code class="docutils literal"><span class="pre">Aligner</span></code> classes. The main
methods provided by <code class="docutils literal"><span class="pre">Aligners</span></code> are the <code class="docutils literal"><span class="pre">check_dependencies</span></code> and
<code class="docutils literal"><span class="pre">identify_insertions</span></code> methods. The former is used to check whether
required external dependencies are present in <code class="docutils literal"><span class="pre">$PATH</span></code>, whilst the latter
should be called with a reference instance and the sequence read files
to identify insertions. The implementation of the <code class="docutils literal"><span class="pre">identify_insertions</span></code>
method is specific to each aligner and should be overridden in each
subclass.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>reference</strong> (<a class="reference internal" href="#imfusion.build.indexers.Reference" title="imfusion.build.indexers.Reference"><em>Reference</em></a>) &#8211; Reference class describing the reference paths.</li>
<li><strong>logger</strong> (<em>logging.Logger</em>) &#8211; Logger to be used for logging messages.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="imfusion.insertions.aligners.Aligner.check_dependencies">
<code class="descname">check_dependencies</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/base.html#Aligner.check_dependencies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.Aligner.check_dependencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if all required external dependencies are in <code class="docutils literal"><span class="pre">$PATH</span></code>.</p>
<p>Raises a ValueError if any dependencies are missing.</p>
</dd></dl>

<dl class="classmethod">
<dt id="imfusion.insertions.aligners.Aligner.configure_args">
<em class="property">classmethod </em><code class="descname">configure_args</code><span class="sig-paren">(</span><em>parser</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/base.html#Aligner.configure_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.Aligner.configure_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Configures an argument parser for the Indexer.</p>
<p>Used by <code class="docutils literal"><span class="pre">imfusion-build</span></code> to configure the sub-command for
this indexer (if registered as an Indexer using the
<code class="docutils literal"><span class="pre">register_indexer</span></code> function).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>parser</strong> (<em>argparse.ArgumentParser</em>) &#8211; Argument parser to configure.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.insertions.aligners.Aligner.dependencies">
<code class="descname">dependencies</code><a class="headerlink" href="#imfusion.insertions.aligners.Aligner.dependencies" title="Permalink to this definition">¶</a></dt>
<dd><p>External dependencies required by aligner.</p>
</dd></dl>

<dl class="classmethod">
<dt id="imfusion.insertions.aligners.Aligner.from_args">
<em class="property">classmethod </em><code class="descname">from_args</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/base.html#Aligner.from_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.Aligner.from_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs an Indexer instance from given arguments.</p>
<p>Instantiates an Indexer instance from the given argparse arguments.
Uses the <code class="docutils literal"><span class="pre">_parse_args</span></code> method internally, which performs the actual
argument-to-parameter extraction. As such, the <code class="docutils literal"><span class="pre">_parse_args</span></code> method
should be overridden in any subclasses with extra parameters.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<em>argparse.Namespace</em>) &#8211; Arguments to parse.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="imfusion.insertions.aligners.Aligner.identify_insertions">
<code class="descname">identify_insertions</code><span class="sig-paren">(</span><em>fastq_path</em>, <em>output_dir</em>, <em>fastq2_path=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/base.html#Aligner.identify_insertions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.Aligner.identify_insertions" title="Permalink to this definition">¶</a></dt>
<dd><p>Identifies insertions from given reads.</p>
<p>Aligns RNA-seq reads to the reference genome and uses this alignment
to identify gene-transposon fusions. These gene-transposon fusions
are summarized to transposon insertions and returned.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>fastq_path</strong> (<em>Path</em>) &#8211; Path to fastq file containing sequence reads. For paired-end data,
this should refer to the first read of the pair.</li>
<li><strong>output_dir</strong> (<em>Path</em>) &#8211; Output directory, to use for output files such as the alignment.</li>
<li><strong>fastq2_path</strong> (<em>Path</em>) &#8211; For paired-end sequencing data, path to fastq file containing
the second read of the pair.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Yields:</th><td class="field-body"><p class="first last"><em>Insertion</em> &#8211; Yields the identified insertions.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="imfusion.insertions.aligners.StarAligner">
<em class="property">class </em><code class="descclassname">imfusion.insertions.aligners.</code><code class="descname">StarAligner</code><span class="sig-paren">(</span><em>reference</em>, <em>assemble=False</em>, <em>assemble_args=None</em>, <em>min_flank=12</em>, <em>threads=1</em>, <em>extra_args=None</em>, <em>logger=None</em>, <em>filter_features=True</em>, <em>filter_orientation=True</em>, <em>filter_blacklist=None</em>, <em>external_sort=False</em>, <em>merge_junction_dist=10</em>, <em>max_spanning_dist=300</em>, <em>max_junction_dist=10000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/star.html#StarAligner"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.StarAligner" title="Permalink to this definition">¶</a></dt>
<dd><p>STAR aligner.</p>
<p>Aligner that identifies insertions using STAR, by aligning reads using STAR
(including chimeric alignments) and identifying gene-transposon fusions
from chimeric read alignments.</p>
<p>As STAR provides chimeric alignment information but does not provide a
an actual list of fusions, we identify fusions by summarizing these
alignments ourselves. The strategy to do so is as follows:</p>
<blockquote>
<div><ol class="arabic simple">
<li>We filter for chimeric reads involving the transposon and genomic
loci, to reduce the number of reads we need to consider.</li>
<li>We select chimeric reads (or mates) that actually overlap the
breakpoint of the fusion, as these alignments provide the exact
position of the fusion. These reads are grouped by their positions
to determine the number of reads that support each fusion. Fusions
with close distance (on both the genomic and transposon loci) are
merged to account for slight variations in the alignment.</li>
<li><dl class="first docutils">
<dt>For paired-end data:</dt>
<dd><ul class="first last">
<li>We select mates that span the fusion site
(meaning that they do not overlap the breakpoint, therefore we
only have an approximate position) and group mates of which both
ends are with close proximity of eachother.</li>
<li>We then try to assign these groups of spanning mates to
previously identified &#8216;junction&#8217; fusions that are in the close
vicinity. If we find a match, the number of mates in the group is
registered as &#8216;spanning support&#8217; for that fusion. If no match is
found, we consider the group to identify a unique fusion and
derive an approximate location for the fusion from the
mate locations.</li>
</ul>
</dd>
</dl>
</li>
<li>We return the combined list of fusions. For single-end data, this
only contains junction fusions, whilst for paired-end data this is
the combined list of junction and unassigned spanning fusions.</li>
</ol>
</div></blockquote>
<p>To avoid memory issues with STAR during the sorting of bam files, sorting
can be performed using Sambamba. Other external dependencies of this
aligner are STAR itself and Stringtie, which is used for the assembly of
novel transcripts (if requested).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>reference</strong> (<a class="reference internal" href="#imfusion.build.indexers.Reference" title="imfusion.build.indexers.Reference"><em>Reference</em></a>) &#8211; Reference class describing the reference paths.</li>
<li><strong>assemble</strong> (<em>bool</em>) &#8211; Whether to perform the assembly of novel transcripts using Stringie.</li>
<li><strong>assemble_args</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em><em></em>) &#8211; Extra arguments to pass to Stringie during assembly.</li>
<li><strong>min_flank</strong> (<em>int</em>) &#8211; Minimum number of flanking bases required on both sides of the
breakpoint for a fusion to be considered valid. This means that, for
a split read overlapping the fusion boundary, at least <code class="docutils literal"><span class="pre">min_flank</span></code>
bases should be on either sides of the fusion.</li>
<li><strong>threads</strong> (<em>int</em>) &#8211; Number of threads to use.</li>
<li><strong>extra_args</strong> &#8211; Extra arguments to pass to STAR for alignment.</li>
<li><strong>logger</strong> (<em>logging.Logger</em>) &#8211; Logger to be used for logging messages.</li>
<li><strong>filter_features</strong> (<em>bool</em>) &#8211; Whether insertions should be filtered to remove non SA/SD insertions.</li>
<li><strong>filter_orientation</strong> (<em>bool</em>) &#8211; Whether insertions should be filtered for fusions in which the
transposon feature is in the wrong orientation.</li>
<li><strong>filter_blacklist</strong> (<em>List</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; List of gene ids to filter insertions for.</li>
<li><strong>external_sort</strong> (<em>bool</em>) &#8211; Whether sorting should be performed using STAR (False) or using
Sambamba (True). Sambamba uses less memory than STAR, but is slower.</li>
<li><strong>merge_junction_dist</strong> (<em>int</em>) &#8211; Maximum distance within which fusions supported by split reads
(overlapping the junction, so that the exact breakpoint is known) are
merged. This merging avoids calling multiple fusions due to slight
variations in the alignment, although this value should not be chosen
too large to avoid merging distinct insertions.</li>
<li><strong>max_spanning_dist</strong> (<em>int</em>) &#8211; Maximum distance within which spanning mate pairs (mates that do not
overlap the fusion junction) are grouped when summarizing spanning
chimeric reads. Both mates from two pairs need to be within this
distance of each other to be merged. The value should be chosen to
reflect the expected or emprical insert size.</li>
<li><strong>max_junction_dist</strong> (<em>int</em>) &#8211; Maxixmum distance within which groups of spanning mates are assigned
to a junction fusion (which is supported by split reads, so that its
exact position is known). Groups that cannot be assigned to a junction
fusion are considered to arise from a separate insertion.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="classmethod">
<dt id="imfusion.insertions.aligners.StarAligner.configure_args">
<em class="property">classmethod </em><code class="descname">configure_args</code><span class="sig-paren">(</span><em>parser</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/star.html#StarAligner.configure_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.StarAligner.configure_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Configures an argument parser for the Indexer.</p>
<p>Used by <code class="docutils literal"><span class="pre">imfusion-build</span></code> to configure the sub-command for
this indexer (if registered as an Indexer using the
<code class="docutils literal"><span class="pre">register_indexer</span></code> function).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>parser</strong> (<em>argparse.ArgumentParser</em>) &#8211; Argument parser to configure.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="imfusion.insertions.aligners.StarAligner.dependencies">
<code class="descname">dependencies</code><a class="headerlink" href="#imfusion.insertions.aligners.StarAligner.dependencies" title="Permalink to this definition">¶</a></dt>
<dd><p>External dependencies required by aligner.</p>
</dd></dl>

<dl class="method">
<dt id="imfusion.insertions.aligners.StarAligner.identify_insertions">
<code class="descname">identify_insertions</code><span class="sig-paren">(</span><em>fastq_path</em>, <em>output_dir</em>, <em>fastq2_path=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/star.html#StarAligner.identify_insertions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.StarAligner.identify_insertions" title="Permalink to this definition">¶</a></dt>
<dd><p>Identifies insertions from given reads.</p>
<p>Aligns RNA-seq reads to the reference genome and uses this alignment
to identify gene-transposon fusions. These gene-transposon fusions
are summarized to transposon insertions and returned.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>fastq_path</strong> (<em>Path</em>) &#8211; Path to fastq file containing sequence reads. For paired-end data,
this should refer to the first read of the pair.</li>
<li><strong>output_dir</strong> (<em>Path</em>) &#8211; Output directory, to use for output files such as the alignment.</li>
<li><strong>fastq2_path</strong> (<em>Path</em>) &#8211; For paired-end sequencing data, path to fastq file containing
the second read of the pair.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Yields:</th><td class="field-body"><p class="first last"><em>Insertion</em> &#8211; Yields the identified insertions.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="imfusion.insertions.aligners.TophatAligner">
<em class="property">class </em><code class="descclassname">imfusion.insertions.aligners.</code><code class="descname">TophatAligner</code><span class="sig-paren">(</span><em>reference</em>, <em>assemble=False</em>, <em>assemble_args=None</em>, <em>min_flank=12</em>, <em>threads=1</em>, <em>extra_args=None</em>, <em>logger=None</em>, <em>filter_features=True</em>, <em>filter_orientation=True</em>, <em>filter_blacklist=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/tophat.html#TophatAligner"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.TophatAligner" title="Permalink to this definition">¶</a></dt>
<dd><p>Tophat2 aligner.</p>
<p>Aligner that identifies insertions using Tophat-Fusion (as implemented in
Tophat2), which aligns RNA-seq reads and returns a summary file detailing
the identified fusions. Gene-transposon fusions are identified directly
from this summary file.</p>
<p>External dependencies include Tophat2 and Bowtie1, which are both used for
alignment. Stringtie is required for the detection of novel transcripts.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>reference</strong> (<a class="reference internal" href="#imfusion.build.indexers.Reference" title="imfusion.build.indexers.Reference"><em>Reference</em></a>) &#8211; Reference class describing the reference paths.</li>
<li><strong>assemble</strong> (<em>bool</em>) &#8211; Whether to perform the assembly of novel transcripts using Stringie.</li>
<li><strong>assemble_args</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em><em></em>) &#8211; Extra arguments to pass to Stringie during assembly.</li>
<li><strong>min_flank</strong> (<em>int</em>) &#8211; Minimum number of flanking bases required on both sides of the
breakpoint for a fusion to be considered valid. This means that, for
a split read overlapping the fusion boundary, at least <code class="docutils literal"><span class="pre">min_flank</span></code>
bases should be on either sides of the fusion.</li>
<li><strong>threads</strong> (<em>int</em>) &#8211; Number of threads to use.</li>
<li><strong>extra_args</strong> &#8211; Extra arguments to pass to Tophat2 for alignment.</li>
<li><strong>logger</strong> (<em>logging.Logger</em>) &#8211; Logger to be used for logging messages.</li>
<li><strong>filter_features</strong> (<em>bool</em>) &#8211; Whether insertions should be filtered to remove non SA/SD insertions.</li>
<li><strong>filter_orientation</strong> (<em>bool</em>) &#8211; Whether insertions should be filtered for fusions in which the
transposon feature is in the wrong orientation.</li>
<li><strong>filter_blacklist</strong> (<em>List</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; List of gene ids to filter insertions for.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="imfusion.insertions.aligners.TophatAligner.identify_insertions">
<code class="descname">identify_insertions</code><span class="sig-paren">(</span><em>fastq_path</em>, <em>output_dir</em>, <em>fastq2_path=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/insertions/aligners/tophat.html#TophatAligner.identify_insertions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.insertions.aligners.TophatAligner.identify_insertions" title="Permalink to this definition">¶</a></dt>
<dd><p>Identifies insertions from given reads.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="differential-expression">
<h2>Differential expression<a class="headerlink" href="#differential-expression" title="Permalink to this headline">¶</a></h2>
<div class="section" id="quantification">
<h3>Quantification<a class="headerlink" href="#quantification" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">imfusion.expression</span></code> module provides several functions for generating,
reading and normalizing exon expression counts. These functions are used by
IM-Fusion to prepare counts for the differential expression tests.</p>
<dl class="function">
<dt id="imfusion.expression.generate_exon_counts">
<code class="descclassname">imfusion.expression.</code><code class="descname">generate_exon_counts</code><span class="sig-paren">(</span><em>bam_files</em>, <em>gtf_path</em>, <em>names=None</em>, <em>extra_kws=None</em>, <em>tmp_dir=None</em>, <em>keep_tmp=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/counts.html#generate_exon_counts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.generate_exon_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates exon counts for given bam files using featureCounts.</p>
<p>This function is used to generate a m-by-n matrix (m = number of samples,
n = number of exons) of exon expression counts. This matrix is generated
using featureCounts, whose results are then parsed and returned.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>bam_files</strong> (<em>list</em><em>[</em><em>pathlib.Path</em><em>]</em><em></em>) &#8211; List of paths to the bam files for which counts should be generated.</li>
<li><strong>gtf_path</strong> (<em>pathlib.Path</em>) &#8211; Path to the GTF file containing gene features.</li>
<li><strong>names</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em><em></em>) &#8211; Alternative names to use for the given bam
files. Keys of the dict should correspond to bam file paths, values
should reflect the sample names that should be used in the
resulting count matrix.</li>
<li><strong>extra_kws</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>tuple</em><em>]</em><em>:</em>) &#8211; Dictionary of extra arguments that should be passed to feature counts.
Keys should correspond to argument names (including dashes),
values should be tuples containing the argument values.</li>
<li><strong>tmp_dir</strong> (<em>pathlib.Path</em>) &#8211; Temp directory to use for generating counts.</li>
<li><strong>keep_tmp</strong> (<em>bool</em>) &#8211; Whether to keep the temp directory (default = False).</li>
<li><strong>**kwargs</strong> &#8211; Any kwargs are passed to <cite>feature_counts</cite>.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">DataFrame containing counts. The index of the DataFrame contains
gene ids corresponding to exons in the gff file, the columns
correspond to samples/bam files. Column names are either the bam
file paths, or the alternative sample names if given.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pandas.DataFrame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="imfusion.expression.read_exon_counts">
<code class="descclassname">imfusion.expression.</code><code class="descname">read_exon_counts</code><span class="sig-paren">(</span><em>file_path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/counts.html#read_exon_counts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.read_exon_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads exon counts from an IM-Fusion expression file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>file_path</strong> (<em>pathlib.Path</em>) &#8211; Path to the exon count file.</li>
<li><strong>gene_id</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; Optional gene_id.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Matrix of exon counts. The rows correspond to the
counted features, the columns correspond to the index values
(chomosome, position etc.) and the samples.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pd.DataFrame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="imfusion.expression.normalize_counts">
<code class="descclassname">imfusion.expression.</code><code class="descname">normalize_counts</code><span class="sig-paren">(</span><em>counts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/counts.html#normalize_counts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.normalize_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalizes counts for sequencing depth using median-of-ratios.</p>
<p>Normalizes gene expression counts for differences in sequencing depth
between samples using the median-of-ratios approach described in DESeq2.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>counts</strong> (<em>pd.DataFrame</em>) &#8211; Matrix of gene counts, with genes along the rows and samples
along the columns.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Matrix of normalized expression counts.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">pd.DataFrame</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="imfusion.expression.estimate_size_factors">
<code class="descclassname">imfusion.expression.</code><code class="descname">estimate_size_factors</code><span class="sig-paren">(</span><em>counts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/counts.html#estimate_size_factors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.estimate_size_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates size factors using median-of-ratios.</p>
<p>Calculates normalization factors for the median-of-ratios approach for
normalizing expression counts for differences in sequencing depths between
samples.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>counts</strong> (<em>pd.DataFrame</em>) &#8211; Matrix of gene counts, with genes along the rows and samples
along the columns.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Array of normalization factors, with one entry per sample.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">np.Array</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="de-tests">
<h3>DE tests<a class="headerlink" href="#de-tests" title="Permalink to this headline">¶</a></h3>
<p>IM-Fusion provides three differential expression tests, which are implemented
in their own functions (<code class="docutils literal"><span class="pre">test_de_exon</span></code>, <code class="docutils literal"><span class="pre">test_de_exon_single</span></code> and
<code class="docutils literal"><span class="pre">test_de_gene</span></code>). Each of these functions returns a corresponding test result
instance (see below for more details). The <code class="docutils literal"><span class="pre">test_de</span></code> function is a
convenient wrapper that applies the group-wise exon-level test for multiple
genes, and provides an optional fallback to the gene-level test for genes that
cannot be tested at the exon-level.</p>
<dl class="function">
<dt id="imfusion.expression.test_de">
<code class="descclassname">imfusion.expression.</code><code class="descname">test_de</code><span class="sig-paren">(</span><em>insertions</em>, <em>exon_counts</em>, <em>gene_ids</em>, <em>fallback_to_gene=False</em>, <em>gene_counts=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#test_de"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test_de" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs and summarizes group-wise DE tests for multiple genes.</p>
<p>High-level function that performs the group-wise DE tests for multiple
genes and summarizes the results in a single table. Provides an optional
fallback to perform the gene-level test for genes that can not be tested
using the exon level test (typically due to a lack of exons before/after
the insertion sites in the gene).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>insertions</strong> (<em>List</em><em>[</em><em>Insertion</em><em>]</em><em></em>) &#8211; Insertions to use for test.</li>
<li><strong>counts</strong> (<em>pd.DataFrame</em>) &#8211; Exon expression counts to use for test. Expected to conform to
the format returned by the <cite>read_exon_counts</cite> function.</li>
<li><strong>gene_ids</strong> (<em>List</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; IDs of genes to test for differential expression. Expected to conform
with gene_ids used for the insertions and the expression counts.</li>
<li><strong>fallback_to_gene</strong> (<em>bool</em>) &#8211; If true, the gene-level test is used as a fallback for genes that do
not have any exons before/after their insertion sites. If false, no
test is performed for these genes.</li>
<li><strong>gene_counts</strong> (<em>pd.DataFrame</em>) &#8211; Optional matrix of gene expression counts to use for the gene-level
test. If not given, the exon expression counts are used to estimate
gene-level expression by summing the expression values of the gene
exons.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">DataFrame summarizing the results of the DE tests. Contains three
columns: &#8216;gene_id&#8217;, &#8216;p_value&#8217; and &#8216;direction&#8217;. The p-value indicates
the significance of the differential expression for the given gene,
whilst the direction indicates whether the expression goes up (=1)
in samples with an insertion or goes down (=-1).</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pd.DataFrame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="imfusion.expression.test_de_exon">
<code class="descclassname">imfusion.expression.</code><code class="descname">test_de_exon</code><span class="sig-paren">(</span><em>insertions</em>, <em>exon_counts</em>, <em>gene_id</em>, <em>pos_samples=None</em>, <em>neg_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#test_de_exon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test_de_exon" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs the groupwise exon-level differential expression test.</p>
<p>Tests if the expression of exons after the insertion site(s) in a
gene is significantly increased or decreased in samples with an
insertion (pos_samples) compared to samples without an insertion
(neg_samples). The test is performed by comparing normalized counts
after the insertion sites between samples with and without an insertion
in the gene, using the non-parametric Mann-Whitney-U test.</p>
<p>Note that the before/after split for the groupwise test is taken as the
common set of before/after exons over all samples with an insertion. In
cases where either set is empty, for example due to insertions before the
first exon of the gene, we attempt to drop samples that prevent a proper
split and perform the test without these samples.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>insertions</strong> (<em>List</em><em>[</em><em>Insertion</em><em>] or </em><em>pd.DataFrame</em>) &#8211; List of insertions.</li>
<li><strong>exon_counts</strong> (<em>pandas.DataFrame</em>) &#8211; Matrix containing exon counts, with samples along the columns and
exons along the rows. The DataFrame should have a multi-index
containing the chromosome, start, end and strand of the exon. The
samples should correspond with samples in the insertions frame.</li>
<li><strong>gene_id</strong> (<em>str</em>) &#8211; ID of the gene of interest. Should correspond with a
gene in the count matrix.</li>
<li><strong>pos_samples</strong> (<em>Set</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; Set of positive samples (with insertion) to use in the test. Defaults
to all samples with an insertion in the gene of interest.</li>
<li><strong>neg_samples</strong> (<em>Set</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; Set of negative samples (without insertion) to use in the test.
Defaults to all samples not in the positive set.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Result of the differential expression test.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#imfusion.expression.test.DeResult" title="imfusion.expression.test.DeResult">DeResult</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="imfusion.expression.test_de_exon_single">
<code class="descclassname">imfusion.expression.</code><code class="descname">test_de_exon_single</code><span class="sig-paren">(</span><em>insertions</em>, <em>exon_counts</em>, <em>insertion_id</em>, <em>gene_id</em>, <em>neg_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#test_de_exon_single"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test_de_exon_single" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs the single sample exon-level differential expression test.</p>
<p>Tests if the expression of exons after the insertion site(s) in a
gene is significantly increased or decreased in the sample with the given
insertion compared to samples without an insertion (neg_samples).
The test is performed by comparing normalized counts between the given
sample and the set of background samples using the negative binomial
distribution.</p>
<p>Note that the before/after split for the groupwise test is taken as the
common set of before/after exons over all samples with an insertion. In
cases where either set is empty, for example due to insertions before the
first exon of the gene, we attempt to drop samples that prevent a proper
split and perform the test without these samples.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>insertions</strong> (<em>Union</em><em>[</em><em>List</em><em>[</em><em>Insertion</em><em>]</em><em></em><em>, </em><em>pd.DataFrame</em><em>]</em><em></em>) &#8211; List of insertions.</li>
<li><strong>exon_counts</strong> (<em>pandas.DataFrame</em>) &#8211; Matrix containing exon counts, with samples along the columns and
exons along the rows. The DataFrame should have a multi-index
containing the chromosome, start, end and strand of the exon. The
samples should correspond with samples in the insertions frame.</li>
<li><strong>gene_id</strong> (<em>str</em>) &#8211; ID of the gene of interest. Should correspond with a
gene in the count matrix.</li>
<li><strong>neg_samples</strong> (<em>Set</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; Set of negative samples (without insertion) to use in the test.
Defaults to all samples without an insertion in the gene.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Result of the differential expression test.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#imfusion.expression.test.DeSingleResult" title="imfusion.expression.test.DeSingleResult">DeSingleResult</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="imfusion.expression.test_de_gene">
<code class="descclassname">imfusion.expression.</code><code class="descname">test_de_gene</code><span class="sig-paren">(</span><em>insertions</em>, <em>gene_counts</em>, <em>gene_id</em>, <em>pos_samples=None</em>, <em>neg_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#test_de_gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test_de_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs the groupwise gene-level differential expression test.</p>
<p>Tests if the expression of the given gene is signficantly increased
or decreased in samples with an insertion in the gene. Significance
is calculated using a mannwhitneyu test between the two groups, after
normalizing for sequencing depth using median-of-ratios normalization
(as implemented in DESeq2).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>insertions</strong> (<em>Union</em><em>[</em><em>List</em><em>[</em><em>Insertion</em><em>]</em><em></em><em>, </em><em>pd.DataFrame</em><em>]</em><em></em>) &#8211; List of insertions.</li>
<li><strong>gene_counts</strong> (<em>pandas.DataFrame</em>) &#8211; Matrix containing gene counts, with samples along the columns and
genes along the rows.</li>
<li><strong>gene_id</strong> (<em>str</em>) &#8211; ID of the gene of interest. Should correspond with a
gene in the count matrix.</li>
<li><strong>pos_samples</strong> (<em>Set</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; Set of positive samples (with insertion) to use in the test. Defaults
to all samples with an insertion in the gene of interest.</li>
<li><strong>neg_samples</strong> (<em>Set</em><em>[</em><em>str</em><em>]</em><em></em>) &#8211; Set of negative samples (without insertion) to use in the test.
Defaults to all samples not in the positive set.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Result of the differential expression test.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#imfusion.expression.test.DeResult" title="imfusion.expression.test.DeResult">DeResult</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="de-results">
<h3>DE results<a class="headerlink" href="#de-results" title="Permalink to this headline">¶</a></h3>
<p>Each of the three differential expression tests returns a result instance that
contains the results of the test and provides several functions for
visualizing the test result.</p>
<dl class="class">
<dt id="imfusion.expression.test.DeResult">
<em class="property">class </em><code class="descclassname">imfusion.expression.test.</code><code class="descname">DeResult</code><span class="sig-paren">(</span><em>before</em>, <em>after</em>, <em>pos_samples</em>, <em>neg_samples</em>, <em>dropped_samples</em>, <em>direction</em>, <em>p_value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#DeResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test.DeResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Class embodying the results of the groupwise exon-level DE test.</p>
<dl class="attribute">
<dt id="DeResult.before">
<code class="descname">before</code><a class="headerlink" href="#DeResult.before" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.DataFrame</em> &#8211; Expression counts of exons before the split.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeResult.after">
<code class="descname">after</code><a class="headerlink" href="#DeResult.after" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.DataFrame</em> &#8211; Expression counts of exons after the split.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeResult.positive_samples">
<code class="descname">positive_samples</code><a class="headerlink" href="#DeResult.positive_samples" title="Permalink to this definition">¶</a></dt>
<dd><p><em>Set[str]</em> &#8211; Samples with an insertion.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeResult.negative_samples">
<code class="descname">negative_samples</code><a class="headerlink" href="#DeResult.negative_samples" title="Permalink to this definition">¶</a></dt>
<dd><p><em>Set[str]</em> &#8211; Samples without an insertion.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeResult.dropped_samples">
<code class="descname">dropped_samples</code><a class="headerlink" href="#DeResult.dropped_samples" title="Permalink to this definition">¶</a></dt>
<dd><p><em>Set[str]</em> &#8211; Samples that were dropped from the analysis, because of violating
the constraints on the minimum number of exons before/after the split.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeResult.direction">
<code class="descname">direction</code><a class="headerlink" href="#DeResult.direction" title="Permalink to this definition">¶</a></dt>
<dd><p><em>int</em> &#8211; Direction of the differential expression (1 = positive, -1 = negative).</p>
</dd></dl>

<dl class="attribute">
<dt id="DeResult.p_value">
<code class="descname">p_value</code><a class="headerlink" href="#DeResult.p_value" title="Permalink to this definition">¶</a></dt>
<dd><p><em>float</em> &#8211; P-value of the differential expression test.</p>
</dd></dl>

<dl class="method">
<dt id="imfusion.expression.test.DeResult.plot_boxplot">
<code class="descname">plot_boxplot</code><span class="sig-paren">(</span><em>log=False</em>, <em>ax=None</em>, <em>show_points=True</em>, <em>box_kws=None</em>, <em>strip_kws=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#DeResult.plot_boxplot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test.DeResult.plot_boxplot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots boxplot of &#8216;after&#8217; expression for samples with/without
insertions in the gene.</p>
</dd></dl>

<dl class="method">
<dt id="imfusion.expression.test.DeResult.plot_sums">
<code class="descname">plot_sums</code><span class="sig-paren">(</span><em>log=False</em>, <em>ax=None</em>, <em>line_kws=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#DeResult.plot_sums"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test.DeResult.plot_sums" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the distribution of before/after counts for the samples.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="imfusion.expression.test.DeSingleResult">
<em class="property">class </em><code class="descclassname">imfusion.expression.test.</code><code class="descname">DeSingleResult</code><span class="sig-paren">(</span><em>before</em>, <em>after</em>, <em>pos_sample</em>, <em>neg_samples</em>, <em>direction</em>, <em>p_value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#DeSingleResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test.DeSingleResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Class embodying the results of the single-sample exon-level DE test.</p>
<dl class="attribute">
<dt id="DeSingleResult.before">
<code class="descname">before</code><a class="headerlink" href="#DeSingleResult.before" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.DataFrame</em> &#8211; Expression counts of exons before the split.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeSingleResult.after">
<code class="descname">after</code><a class="headerlink" href="#DeSingleResult.after" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.DataFrame</em> &#8211; Expression counts of exons after the split.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeSingleResult.positive_sample">
<code class="descname">positive_sample</code><a class="headerlink" href="#DeSingleResult.positive_sample" title="Permalink to this definition">¶</a></dt>
<dd><p><em>str</em> &#8211; Samples with an insertion.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeSingleResult.negative_samples">
<code class="descname">negative_samples</code><a class="headerlink" href="#DeSingleResult.negative_samples" title="Permalink to this definition">¶</a></dt>
<dd><p><em>Set[str]</em> &#8211; Samples without an insertion.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeSingleResult.direction">
<code class="descname">direction</code><a class="headerlink" href="#DeSingleResult.direction" title="Permalink to this definition">¶</a></dt>
<dd><p><em>int</em> &#8211; Direction of the differential expression (1 = positive, -1 = negative).</p>
</dd></dl>

<dl class="attribute">
<dt id="DeSingleResult.p_value">
<code class="descname">p_value</code><a class="headerlink" href="#DeSingleResult.p_value" title="Permalink to this definition">¶</a></dt>
<dd><p><em>float</em> &#8211; P-value of the differential expression test.</p>
</dd></dl>

<dl class="method">
<dt id="imfusion.expression.test.DeSingleResult.plot_sums">
<code class="descname">plot_sums</code><span class="sig-paren">(</span><em>log=False</em>, <em>ax=None</em>, <em>line_kws=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#DeSingleResult.plot_sums"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test.DeSingleResult.plot_sums" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the distribution of before/after counts for the samples.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="imfusion.expression.test.DeGeneResult">
<em class="property">class </em><code class="descclassname">imfusion.expression.test.</code><code class="descname">DeGeneResult</code><span class="sig-paren">(</span><em>counts</em>, <em>pos_samples</em>, <em>neg_samples</em>, <em>direction</em>, <em>p_value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#DeGeneResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test.DeGeneResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Class embodying the results of the groupwise gene-level DE test.</p>
<dl class="attribute">
<dt id="DeGeneResult.counts">
<code class="descname">counts</code><a class="headerlink" href="#DeGeneResult.counts" title="Permalink to this definition">¶</a></dt>
<dd><p><em>pandas.Series</em> &#8211; Expression counts of gene for the different samples.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeGeneResult.positive_samples">
<code class="descname">positive_samples</code><a class="headerlink" href="#DeGeneResult.positive_samples" title="Permalink to this definition">¶</a></dt>
<dd><p><em>Set[str]</em> &#8211; Samples with an insertion.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeGeneResult.negative_samples">
<code class="descname">negative_samples</code><a class="headerlink" href="#DeGeneResult.negative_samples" title="Permalink to this definition">¶</a></dt>
<dd><p><em>Set[str]</em> &#8211; Samples without an insertion.</p>
</dd></dl>

<dl class="attribute">
<dt id="DeGeneResult.direction">
<code class="descname">direction</code><a class="headerlink" href="#DeGeneResult.direction" title="Permalink to this definition">¶</a></dt>
<dd><p><em>int</em> &#8211; Direction of the differential expression (1 = positive, -1 = negative).</p>
</dd></dl>

<dl class="attribute">
<dt id="DeGeneResult.p_value">
<code class="descname">p_value</code><a class="headerlink" href="#DeGeneResult.p_value" title="Permalink to this definition">¶</a></dt>
<dd><p><em>float</em> &#8211; P-value of the differential expression test.</p>
</dd></dl>

<dl class="method">
<dt id="imfusion.expression.test.DeGeneResult.plot_boxplot">
<code class="descname">plot_boxplot</code><span class="sig-paren">(</span><em>ax=None</em>, <em>log=False</em>, <em>box_kws=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/imfusion/expression/test.html#DeGeneResult.plot_boxplot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#imfusion.expression.test.DeGeneResult.plot_boxplot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots boxplot comparing expression between the two groups.</p>
</dd></dl>

</dd></dl>

</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="extras.html" class="btn btn-neutral float-right" title="Extras" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="usage.html" class="btn btn-neutral" title="Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Julian de Ruiter.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.3.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>