

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Installation &mdash; IM-Fusion 0.3.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="IM-Fusion 0.3.1 documentation" href="index.html"/>
        <link rel="next" title="Getting started" href="getting_started.html"/>
        <link rel="prev" title="IM-Fusion" href="index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> IM-Fusion
          

          
          </a>

          
            
            
              <div class="version">
                0.3.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">IM-Fusion</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#using-bioconda-recommended">Using bioconda (recommended)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#using-pip">Using pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="#hardware-dependencies">Hardware dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="#software-dependencies">Software dependencies</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="extras.html">Extras</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">IM-Fusion</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Installation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/installation.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="using-bioconda-recommended">
<h2>Using bioconda (recommended)<a class="headerlink" href="#using-bioconda-recommended" title="Permalink to this headline">¶</a></h2>
<p>The recommended way to install IM-Fusion is using bioconda, as this installs
IM-Fusion together with its external dependencies into an isolated environment
using a single command:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">create</span> <span class="o">-</span><span class="n">n</span> <span class="n">imfusion</span> <span class="n">imfusion</span>
</pre></div>
</div>
<p>This assumes that condas channels have been configured as <a class="reference external" href="https://bioconda.github.io/#set-up-channels">recommended</a>
by bioconda.</p>
<p>Alternatively, IM-Fusion can be installed in an existing environment using:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="n">imfusion</span>
</pre></div>
</div>
<p>Note that the bioconda package does not install rpy2, which is only required
for the single-sample differential expression (DE) test and can be difficult to
install. If you wish to use the single-sample DE test, you can install R and
rpy2 can be installed using as follows:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Using conda</span>
<span class="n">conda</span> <span class="n">install</span> <span class="n">rpy2</span>

<span class="c1"># Using pip</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">rpy2</span>
</pre></div>
</div>
</div>
<div class="section" id="using-pip">
<h2>Using pip<a class="headerlink" href="#using-pip" title="Permalink to this headline">¶</a></h2>
<p>IM-Fusion can also be installed using pip:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">imfusion</span>
</pre></div>
</div>
<p>Note that this does not install any of the required external dependencies,
which must therefore be installed manually (see below for more details).
This does also not install rpy2, which must also be installed using pip if you
plan to use single-sample differential expression test.</p>
<p>If you don&#8217;t already have numpy installed, you may run into the following
error during installation:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">Downloading</span> <span class="n">HTSeq</span><span class="o">-</span><span class="mf">0.7</span><span class="o">.</span><span class="mf">2.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="p">(</span><span class="mi">548</span><span class="n">kB</span><span class="p">)</span>
    <span class="n">Complete</span> <span class="n">output</span> <span class="kn">from</span> <span class="nn">command</span> <span class="n">python</span> <span class="n">setup</span><span class="o">.</span><span class="n">py</span> <span class="n">egg_info</span><span class="p">:</span>
    <span class="n">Setup</span> <span class="n">script</span> <span class="k">for</span> <span class="n">HTSeq</span><span class="p">:</span> <span class="n">Failed</span> <span class="n">to</span> <span class="kn">import</span> <span class="s1">&#39;numpy&#39;</span><span class="o">.</span>
    <span class="n">Please</span> <span class="n">install</span> <span class="n">numpy</span> <span class="ow">and</span> <span class="n">then</span> <span class="k">try</span> <span class="n">again</span> <span class="n">to</span> <span class="n">install</span> <span class="n">HTSeq</span><span class="o">.</span>
</pre></div>
</div>
<p>This can be fixed by first installing numpy before trying to install IM-Fusion,
as numpy is an install-time dependency of one of IM-Fusions dependencies
(HTSeq). Additionally, HTSeq requires liblzma to be installed to build
properly, otherwise you will encounter the following error:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">ld</span><span class="p">:</span> <span class="n">cannot</span> <span class="n">find</span> <span class="o">-</span><span class="n">llzma</span>
<span class="n">collect2</span><span class="p">:</span> <span class="n">error</span><span class="p">:</span> <span class="n">ld</span> <span class="n">returned</span> <span class="mi">1</span> <span class="n">exit</span> <span class="n">status</span>
</pre></div>
</div>
<p>This can be fixed by installing the liblzma-dev package. On linux (assuming an
apt-get based distribution) this can be done as follows:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">sudo</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">update</span>
<span class="n">sudo</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="o">-</span><span class="n">y</span> <span class="n">install</span> <span class="n">liblzma</span><span class="o">-</span><span class="n">dev</span>
</pre></div>
</div>
</div>
<div class="section" id="hardware-dependencies">
<h2>Hardware dependencies<a class="headerlink" href="#hardware-dependencies" title="Permalink to this headline">¶</a></h2>
<p>To build references and align reads using the STAR aligner, you will need to
have ~30GB RAM. For analyses with Tophat2, you will typically need ~5-15GB of
RAM, depending on the number of reads in your sample.</p>
</div>
<div class="section" id="software-dependencies">
<h2>Software dependencies<a class="headerlink" href="#software-dependencies" title="Permalink to this headline">¶</a></h2>
<p>IM-Fusion has been tested on macOS and linux, and is compatible with both
Python 2.7 and Python 3.4+. However, identifying insertions using Tophat-Fusion
only works in Python 2.7, as Tophat2 is unfortunately not compatible with
Python 3.</p>
<p>IM-Fusion has the following external dependencies that must be installed
for full functionality:</p>
<ul class="simple">
<li><a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> - for insertion detection using STAR.</li>
<li><a class="reference external" href="https://ccb.jhu.edu/software/tophat/index.shtml">Tophat2</a>/<a class="reference external" href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a> (Python 2.7 only) &#8211; for insertion detection using
Tophat-Fusion.</li>
<li><a class="reference external" href="http://subread.sourceforge.net">featureCounts</a> &#8211; for generating expression counts.</li>
<li><a class="reference external" href="https://ccb.jhu.edu/software/stringtie">StringTie</a> (optional) - for performing reference-guided transcript assembly.</li>
<li><a class="reference external" href="https://github.com/STAR-Fusion/STAR-Fusion">STAR-Fusion</a> (optional) &#8211; for identifying endogenous gene fusions using STAR.</li>
<li><a class="reference external" href="http://lomereiter.github.io/sambamba">Sambamba</a> (optional) &#8211; for faster sorting of BAM files.</li>
</ul>
<p>Note that either STAR or Tophat2/Bowtie need to be installed (not both
necessarily), depending on your preference for the RNA-seq aligner. If you are
unsure, we recommend using STAR. StringTie is only required if you plan to
perform reference-guided assembly (used to detect novel transcripts resulting
from insertions). Similarly, STAR-Fusion is only needed if you plan to do
endogenous fusion detection. Sambamba is used for faster sorting of BAM files
in the STAR pipeline, but is also not a strict requirement.</p>
<p>Each of these packages can be installed manually by downloading binaries from
the respective websites and making sure the unpacked binaries are available in
the current <code class="docutils literal"><span class="pre">$PATH</span></code>. Alternatively, the dependencies can easily be installed
using bioconda:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">star</span> <span class="n">bowtie</span> <span class="n">tophat</span> <span class="n">stringtie</span> \
    <span class="n">subread</span> <span class="n">star</span><span class="o">-</span><span class="n">fusion</span> <span class="n">sambamba</span>
</pre></div>
</div>
<p>Note that Tophat2 will not function properly in Python 3. Bioconda also does
not provide macOS builds of STAR-Fusion, meaning that STAR-Fusion will have
to be installed manually on macOS.</p>
<p>If you don&#8217;t use conda, a number of these tools can also be installed using
<a class="reference external" href="http://brew.sh">homebrew</a> or <a class="reference external" href="http://linuxbrew.sh">linuxbrew</a>:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">brew</span> <span class="n">tap</span> <span class="n">homebrew</span><span class="o">/</span><span class="n">science</span>
<span class="n">brew</span> <span class="n">update</span>
<span class="n">brew</span> <span class="n">install</span> <span class="n">bowtie</span> <span class="n">star</span> <span class="n">stringtie</span> <span class="n">tophat</span> <span class="n">sambamba</span>
</pre></div>
</div>
<p>Unfortunately, STAR-Fusion and subread (featureCounts) are currently not
available in homebrew and will therefore need to be installed manually.</p>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="getting_started.html" class="btn btn-neutral float-right" title="Getting started" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="IM-Fusion" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Julian de Ruiter.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.3.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>