

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>imfusion.insertions.aligners.tophat &mdash; IM-Fusion 0.3.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../../search.html"/>
    <link rel="top" title="IM-Fusion 0.3.0 documentation" href="../../../../index.html"/>
        <link rel="up" title="Module code" href="../../../index.html"/> 

  
  <script src="../../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../../index.html" class="icon icon-home"> IM-Fusion
          

          
          </a>

          
            
            
              <div class="version">
                0.3.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../index.html">IM-Fusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../extras.html">Extras</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../history.html">History</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">IM-Fusion</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../../index.html">Module code</a> &raquo;</li>
        
      <li>imfusion.insertions.aligners.tophat</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for imfusion.insertions.aligners.tophat</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&quot;&quot;&quot;Implements aligner for building Tophat2 references.&quot;&quot;&quot;</span>

<span class="c1"># pylint: disable=wildcard-import,redefined-builtin,unused-wildcard-import</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">absolute_import</span><span class="p">,</span> <span class="n">division</span><span class="p">,</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">builtins</span> <span class="k">import</span> <span class="o">*</span>
<span class="c1"># pylint: enable=wildcard-import,redefined-builtin,unused-wildcard-import</span>

<span class="kn">from</span> <span class="nn">pathlib2</span> <span class="k">import</span> <span class="n">Path</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">toolz</span>

<span class="kn">from</span> <span class="nn">imfusion.build.indexers.tophat</span> <span class="k">import</span> <span class="n">TophatReference</span>
<span class="kn">from</span> <span class="nn">imfusion.external.stringtie</span> <span class="k">import</span> <span class="n">stringtie_assemble</span>
<span class="kn">from</span> <span class="nn">imfusion.external.tophat</span> <span class="k">import</span> <span class="n">tophat2_align</span>
<span class="kn">from</span> <span class="nn">imfusion.external.util</span> <span class="k">import</span> <span class="n">parse_arguments</span>
<span class="kn">from</span> <span class="nn">imfusion.model</span> <span class="k">import</span> <span class="n">TransposonFusion</span>
<span class="kn">from</span> <span class="nn">imfusion.util</span> <span class="k">import</span> <span class="n">path</span><span class="p">,</span> <span class="n">tabix</span>
<span class="kn">from</span> <span class="nn">imfusion.util.frozendict</span> <span class="k">import</span> <span class="n">frozendict</span>

<span class="kn">from</span> <span class="nn">.base</span> <span class="k">import</span> <span class="n">Aligner</span><span class="p">,</span> <span class="n">register_aligner</span>
<span class="kn">from</span> <span class="nn">..</span> <span class="k">import</span> <span class="n">util</span>


<div class="viewcode-block" id="TophatAligner"><a class="viewcode-back" href="../../../../api.html#imfusion.insertions.aligners.TophatAligner">[docs]</a><span class="k">class</span> <span class="nc">TophatAligner</span><span class="p">(</span><span class="n">Aligner</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Tophat2 aligner.</span>

<span class="sd">    Aligner that identifies insertions using Tophat-Fusion (as implemented in</span>
<span class="sd">    Tophat2), which aligns RNA-seq reads and returns a summary file detailing</span>
<span class="sd">    the identified fusions. Gene-transposon fusions are identified directly</span>
<span class="sd">    from this summary file.</span>

<span class="sd">    External dependencies include Tophat2 and Bowtie1, which are both used for</span>
<span class="sd">    alignment. Stringtie is required for the detection of novel transcripts.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    reference : Reference</span>
<span class="sd">        Reference class describing the reference paths.</span>
<span class="sd">    assemble : bool</span>
<span class="sd">        Whether to perform the assembly of novel transcripts using Stringie.</span>
<span class="sd">    assemble_args : Dict[str, Any]</span>
<span class="sd">        Extra arguments to pass to Stringie during assembly.</span>
<span class="sd">    min_flank : int</span>
<span class="sd">        Minimum number of flanking bases required on both sides of the</span>
<span class="sd">        breakpoint for a fusion to be considered valid. This means that, for</span>
<span class="sd">        a split read overlapping the fusion boundary, at least ``min_flank``</span>
<span class="sd">        bases should be on either sides of the fusion.</span>
<span class="sd">    threads : int</span>
<span class="sd">        Number of threads to use.</span>
<span class="sd">    extra_args :</span>
<span class="sd">        Extra arguments to pass to Tophat2 for alignment.</span>
<span class="sd">    logger : logging.Logger</span>
<span class="sd">        Logger to be used for logging messages.</span>
<span class="sd">    filter_features : bool</span>
<span class="sd">        Whether insertions should be filtered to remove non SA/SD insertions.</span>
<span class="sd">    filter_orientation : bool</span>
<span class="sd">        Whether insertions should be filtered for fusions in which the</span>
<span class="sd">        transposon feature is in the wrong orientation.</span>
<span class="sd">    filter_blacklist : List[str]</span>
<span class="sd">        List of gene ids to filter insertions for.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">reference</span><span class="p">,</span>
                 <span class="n">assemble</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">assemble_args</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">min_flank</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
                 <span class="n">threads</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                 <span class="n">extra_args</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">filter_features</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                 <span class="n">filter_orientation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                 <span class="n">filter_blacklist</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">reference</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_assemble</span> <span class="o">=</span> <span class="n">assemble</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_assemble_args</span> <span class="o">=</span> <span class="n">assemble_args</span> <span class="ow">or</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_min_flank</span> <span class="o">=</span> <span class="n">min_flank</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_threads</span> <span class="o">=</span> <span class="n">threads</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_extra_args</span> <span class="o">=</span> <span class="n">extra_args</span> <span class="ow">or</span> <span class="p">{}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_filter_features</span> <span class="o">=</span> <span class="n">filter_features</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_filter_orientation</span> <span class="o">=</span> <span class="n">filter_orientation</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_filter_blacklist</span> <span class="o">=</span> <span class="n">filter_blacklist</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">dependencies</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">programs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;tophat2&#39;</span><span class="p">,</span> <span class="s1">&#39;bowtie&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_assemble</span><span class="p">:</span>
            <span class="n">programs</span> <span class="o">+=</span> <span class="p">[</span><span class="s1">&#39;stringtie&#39;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">programs</span>

<div class="viewcode-block" id="TophatAligner.identify_insertions"><a class="viewcode-back" href="../../../../api.html#imfusion.insertions.aligners.TophatAligner.identify_insertions">[docs]</a>    <span class="k">def</span> <span class="nf">identify_insertions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fastq_path</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">fastq2_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Identifies insertions from given reads.&quot;&quot;&quot;</span>

        <span class="c1"># Perform alignment using STAR.</span>
        <span class="n">alignment_path</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="s1">&#39;alignment.bam&#39;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">alignment_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Performing alignment using Tophat2&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_align</span><span class="p">(</span><span class="n">fastq_path</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">fastq2_path</span><span class="o">=</span><span class="n">fastq2_path</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Using existing Tophat2 alignment&#39;</span><span class="p">)</span>

        <span class="c1"># Assemble transcripts if requested.</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_assemble</span><span class="p">:</span>
            <span class="n">assembled_path</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="s1">&#39;assembled.gtf.gz&#39;</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">assembled_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Assembling transcripts using Stringie&#39;</span><span class="p">)</span>

                <span class="c1"># Generate assembled GTF.</span>
                <span class="n">stringtie_out_path</span> <span class="o">=</span> <span class="n">assembled_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
                <span class="n">stringtie_assemble</span><span class="p">(</span>
                    <span class="n">alignment_path</span><span class="p">,</span>
                    <span class="n">gtf_path</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">gtf_path</span><span class="p">,</span>
                    <span class="n">output_path</span><span class="o">=</span><span class="n">stringtie_out_path</span><span class="p">)</span>

                <span class="c1"># Compress and index.</span>
                <span class="n">tabix</span><span class="o">.</span><span class="n">index_gtf</span><span class="p">(</span><span class="n">stringtie_out_path</span><span class="p">,</span> <span class="n">output_path</span><span class="o">=</span><span class="n">assembled_path</span><span class="p">)</span>
                <span class="n">stringtie_out_path</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Using existing Stringtie assembly&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">assembled_path</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Extract identified fusions.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Extracting gene-transposon fusions&#39;</span><span class="p">)</span>
        <span class="n">fusion_path</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="s1">&#39;fusions.out&#39;</span>
        <span class="n">fusions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_extract_fusions</span><span class="p">(</span><span class="n">fusion_path</span><span class="p">)</span>

        <span class="c1"># Extract insertions.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Summarizing insertions&#39;</span><span class="p">)</span>
        <span class="n">insertions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="n">util</span><span class="o">.</span><span class="n">extract_insertions</span><span class="p">(</span>
                <span class="n">fusions</span><span class="p">,</span>
                <span class="n">gtf_path</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">indexed_gtf_path</span><span class="p">,</span>
                <span class="n">features_path</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">features_path</span><span class="p">,</span>
                <span class="n">assembled_gtf_path</span><span class="o">=</span><span class="n">assembled_path</span><span class="p">,</span>
                <span class="n">ffpm_fastq_path</span><span class="o">=</span><span class="n">fastq_path</span><span class="p">,</span>
                <span class="n">chromosomes</span><span class="o">=</span><span class="kc">None</span><span class="p">))</span>

        <span class="n">insertions</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">filter_insertions</span><span class="p">(</span>
            <span class="n">insertions</span><span class="p">,</span>
            <span class="n">features</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_filter_features</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_filter_orientation</span><span class="p">,</span>
            <span class="n">blacklist</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_filter_blacklist</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">insertion</span> <span class="ow">in</span> <span class="n">insertions</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">insertion</span></div>

    <span class="k">def</span> <span class="nf">_align</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fastq_path</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">fastq2_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="c1"># Setup args.</span>
        <span class="n">transcriptome_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">transcriptome_path</span>

        <span class="n">args</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;--transcriptome-index&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">transcriptome_path</span><span class="p">),</span> <span class="p">),</span>
            <span class="s1">&#39;--fusion-search&#39;</span><span class="p">:</span> <span class="p">(),</span>
            <span class="s1">&#39;--fusion-anchor-length&#39;</span><span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_min_flank</span><span class="p">,</span> <span class="p">),</span>
            <span class="s1">&#39;--bowtie1&#39;</span><span class="p">:</span> <span class="p">(),</span>
            <span class="s1">&#39;--num-threads&#39;</span><span class="p">:</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_threads</span><span class="p">,</span> <span class="p">)</span>
        <span class="p">}</span>

        <span class="n">args</span> <span class="o">=</span> <span class="n">toolz</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_extra_args</span><span class="p">)</span>

        <span class="c1"># Run Tophat2.</span>
        <span class="n">tophat_dir</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="s1">&#39;_tophat&#39;</span>

        <span class="n">tophat2_align</span><span class="p">(</span>
            <span class="n">fastq_path</span><span class="o">=</span><span class="n">fastq_path</span><span class="p">,</span>
            <span class="n">fastq2_path</span><span class="o">=</span><span class="n">fastq2_path</span><span class="p">,</span>
            <span class="n">index_path</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">index_path</span><span class="p">,</span>
            <span class="n">output_dir</span><span class="o">=</span><span class="n">tophat_dir</span><span class="p">,</span>
            <span class="n">extra_args</span><span class="o">=</span><span class="n">args</span><span class="p">)</span>

        <span class="c1"># Symlink alignment into expected location for gene counts.</span>
        <span class="n">path</span><span class="o">.</span><span class="n">symlink_relative</span><span class="p">(</span>
            <span class="n">src_path</span><span class="o">=</span><span class="n">tophat_dir</span> <span class="o">/</span> <span class="s1">&#39;accepted_hits.bam&#39;</span><span class="p">,</span>
            <span class="n">dest_path</span><span class="o">=</span><span class="n">output_dir</span> <span class="o">/</span> <span class="s1">&#39;alignment.bam&#39;</span><span class="p">)</span>

        <span class="n">path</span><span class="o">.</span><span class="n">symlink_relative</span><span class="p">(</span>
            <span class="n">src_path</span><span class="o">=</span><span class="n">tophat_dir</span> <span class="o">/</span> <span class="s1">&#39;fusions.out&#39;</span><span class="p">,</span>
            <span class="n">dest_path</span><span class="o">=</span><span class="n">output_dir</span> <span class="o">/</span> <span class="s1">&#39;fusions.out&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_extract_fusions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fusion_path</span><span class="p">):</span>
        <span class="n">fusion_data</span> <span class="o">=</span> <span class="n">read_fusion_out</span><span class="p">(</span><span class="n">fusion_path</span><span class="p">)</span>

        <span class="n">fusions</span> <span class="o">=</span> <span class="n">extract_transposon_fusions</span><span class="p">(</span>
            <span class="n">fusion_data</span><span class="p">,</span> <span class="n">transposon_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">transposon_name</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">fusion</span> <span class="ow">in</span> <span class="n">fusions</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">fusion</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">configure_args</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">parser</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">configure_args</span><span class="p">(</span><span class="n">parser</span><span class="p">)</span>

        <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Tophat2 arguments&#39;</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--tophat_threads&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--tophat_min_flank&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--tophat_args&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">parse_arguments</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>

        <span class="n">assemble_group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Assembly&#39;</span><span class="p">)</span>
        <span class="n">assemble_group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s1">&#39;--assemble&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_true&#39;</span><span class="p">)</span>

        <span class="n">filt_group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;Filtering&#39;</span><span class="p">)</span>
        <span class="n">filt_group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s1">&#39;--no_filter_orientation&#39;</span><span class="p">,</span>
            <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;filter_orientation&#39;</span><span class="p">,</span>
            <span class="n">default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_false&#39;</span><span class="p">)</span>
        <span class="n">filt_group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s1">&#39;--no_filter_feature&#39;</span><span class="p">,</span>
            <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;filter_features&#39;</span><span class="p">,</span>
            <span class="n">default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_false&#39;</span><span class="p">)</span>
        <span class="n">filt_group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--blacklisted_genes&#39;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">_parse_args</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">args</span><span class="p">):</span>
        <span class="n">kws</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="n">reference</span><span class="o">=</span><span class="n">TophatReference</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">reference</span><span class="p">),</span>
            <span class="n">min_flank</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">tophat_min_flank</span><span class="p">,</span>
            <span class="n">threads</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">tophat_threads</span><span class="p">,</span>
            <span class="n">extra_args</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">tophat_args</span><span class="p">,</span>
            <span class="n">assemble</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">assemble</span><span class="p">,</span>
            <span class="n">filter_features</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">filter_features</span><span class="p">,</span>
            <span class="n">filter_orientation</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">filter_orientation</span><span class="p">,</span>
            <span class="n">filter_blacklist</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">blacklisted_genes</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">toolz</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">_parse_args</span><span class="p">(</span><span class="n">args</span><span class="p">),</span> <span class="n">kws</span><span class="p">)</span></div>


<span class="n">register_aligner</span><span class="p">(</span><span class="s1">&#39;tophat&#39;</span><span class="p">,</span> <span class="n">TophatAligner</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">read_fusion_out</span><span class="p">(</span><span class="n">fusion_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads fusion.out file from Tophat2.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file_path : pathlib.Path</span>
<span class="sd">        Path to the Tophat fusion file (fusions.out).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pandas.DataFrame</span>
<span class="sd">        DataFrame containing gene fusions.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fusion_path</span><span class="p">,</span> <span class="n">Path</span><span class="p">):</span>
        <span class="n">fusion_path</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">fusion_path</span><span class="p">)</span>

    <span class="c1"># Read fusions using pandas.</span>
    <span class="n">names</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;seqnames&#39;</span><span class="p">,</span> <span class="s1">&#39;location_a&#39;</span><span class="p">,</span> <span class="s1">&#39;location_b&#39;</span><span class="p">,</span> <span class="s1">&#39;orientation&#39;</span><span class="p">,</span> <span class="s1">&#39;supp_reads&#39;</span><span class="p">,</span>
        <span class="s1">&#39;supp_mates&#39;</span><span class="p">,</span> <span class="s1">&#39;supp_spanning_mates&#39;</span><span class="p">,</span> <span class="s1">&#39;contradicting_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;flank_a&#39;</span><span class="p">,</span>
        <span class="s1">&#39;flank_b&#39;</span>
    <span class="p">]</span>

    <span class="n">fusions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="n">fusion_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">names</span><span class="o">=</span><span class="n">names</span><span class="p">)</span>

    <span class="c1"># Split combined entries.</span>
    <span class="n">fusions</span> <span class="o">=</span> <span class="n">_split_fields</span><span class="p">(</span><span class="n">fusions</span><span class="p">)</span>

    <span class="c1"># Map orientation to strands.</span>
    <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_a&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">({</span><span class="s1">&#39;f&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">})</span>
    <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_b&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_b&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">({</span><span class="s1">&#39;f&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">})</span>

    <span class="k">return</span> <span class="n">fusions</span>


<span class="k">def</span> <span class="nf">_split_fields</span><span class="p">(</span><span class="n">fusions</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Splits combined seqnames/strand entries in tophat fusion frame.&quot;&quot;&quot;</span>

    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;seqname_a&#39;</span><span class="p">,</span> <span class="s1">&#39;location_a&#39;</span><span class="p">,</span> <span class="s1">&#39;strand_a&#39;</span><span class="p">,</span> <span class="s1">&#39;seqname_b&#39;</span><span class="p">,</span> <span class="s1">&#39;location_b&#39;</span><span class="p">,</span>
        <span class="s1">&#39;strand_b&#39;</span><span class="p">,</span> <span class="s1">&#39;supp_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;supp_mates&#39;</span><span class="p">,</span> <span class="s1">&#39;supp_spanning_mates&#39;</span><span class="p">,</span>
        <span class="s1">&#39;contradicting_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;flank_a&#39;</span><span class="p">,</span> <span class="s1">&#39;flank_b&#39;</span>
    <span class="p">]</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fusions</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Return empty &#39;split&#39; frame.</span>
        <span class="n">fusions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">([],</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Split combined entries.</span>
        <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;seqname_a&#39;</span><span class="p">],</span> <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;seqname_b&#39;</span><span class="p">]</span> <span class="o">=</span> \
            <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;seqnames&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">))</span>

        <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_a&#39;</span><span class="p">],</span> <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_b&#39;</span><span class="p">]</span> <span class="o">=</span> \
            <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;orientation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">list</span><span class="p">))</span>

        <span class="c1"># Subset/Reorder columns.</span>
        <span class="n">fusions</span> <span class="o">=</span> <span class="n">fusions</span><span class="p">[</span><span class="n">columns</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">fusions</span>


<span class="k">def</span> <span class="nf">extract_transposon_fusions</span><span class="p">(</span><span class="n">fusion_data</span><span class="p">,</span> <span class="n">transposon_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts gene-transposon fusions from a Tophat fusion.out file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file_path : pathlib.Path</span>
<span class="sd">        Path to the Tophat fusion file (fusions.out).</span>
<span class="sd">    transposon_name : str</span>
<span class="sd">        Name of the transposon sequence in the augmented reference</span>
<span class="sd">        genome that was used for the alignment.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pandas.DataFrame</span>
<span class="sd">        DataFrame containing gene-transposon fusions.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">is_paired</span> <span class="o">=</span> <span class="p">(</span><span class="nb">any</span><span class="p">(</span><span class="n">fusion_data</span><span class="p">[</span><span class="s1">&#39;supp_mates&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span>
                 <span class="nb">any</span><span class="p">(</span><span class="n">fusion_data</span><span class="p">[</span><span class="s1">&#39;supp_spanning_mates&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">))</span>

    <span class="c1"># Select fusions where one seqname is the transposon and the other isn&#39;t.</span>
    <span class="n">fusion_data</span> <span class="o">=</span> <span class="n">fusion_data</span><span class="o">.</span><span class="n">ix</span><span class="p">[(</span>
        <span class="p">(</span><span class="n">fusion_data</span><span class="p">[</span><span class="s1">&#39;seqname_a&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">transposon_name</span><span class="p">)</span> <span class="o">^</span>
        <span class="p">(</span><span class="n">fusion_data</span><span class="p">[</span><span class="s1">&#39;seqname_b&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">transposon_name</span><span class="p">))]</span>

    <span class="c1"># Build frame with candidate insertions.</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">fusion_row</span> <span class="ow">in</span> <span class="n">fusion_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">yield</span> <span class="n">_to_fusion_obj</span><span class="p">(</span><span class="n">fusion_row</span><span class="p">,</span> <span class="n">transposon_name</span><span class="p">,</span> <span class="n">is_paired</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_to_fusion_obj</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">transposon_name</span><span class="p">,</span> <span class="n">is_paired</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">fusion</span><span class="o">.</span><span class="n">seqname_a</span> <span class="o">==</span> <span class="n">transposon_name</span><span class="p">:</span>
        <span class="n">gen_id</span><span class="p">,</span> <span class="n">tr_id</span> <span class="o">=</span> <span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span>
        <span class="n">gen_dir</span><span class="p">,</span> <span class="n">tr_dir</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gen_id</span><span class="p">,</span> <span class="n">tr_id</span> <span class="o">=</span> <span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span>
        <span class="n">gen_dir</span><span class="p">,</span> <span class="n">tr_dir</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span>

    <span class="n">strand_genome</span> <span class="o">=</span> <span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;strand_&#39;</span> <span class="o">+</span> <span class="n">gen_id</span><span class="p">]</span>
    <span class="n">strand_transposon</span> <span class="o">=</span> <span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;strand_&#39;</span> <span class="o">+</span> <span class="n">tr_id</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">is_paired</span><span class="p">:</span>
        <span class="n">support_junction</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">supp_spanning_mates</span>
        <span class="n">support_spanning</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">supp_mates</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">support_junction</span> <span class="o">=</span> <span class="n">fusion</span><span class="o">.</span><span class="n">supp_reads</span>
        <span class="n">support_spanning</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">return</span> <span class="n">TransposonFusion</span><span class="p">(</span>
        <span class="n">seqname</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;seqname_&#39;</span> <span class="o">+</span> <span class="n">gen_id</span><span class="p">],</span>
        <span class="n">anchor_genome</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;location_&#39;</span> <span class="o">+</span> <span class="n">gen_id</span><span class="p">],</span>
        <span class="n">anchor_transposon</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;location_&#39;</span> <span class="o">+</span> <span class="n">tr_id</span><span class="p">],</span>
        <span class="n">flank_genome</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;flank_&#39;</span> <span class="o">+</span> <span class="n">gen_id</span><span class="p">]</span> <span class="o">*</span> <span class="n">strand_genome</span> <span class="o">*</span> <span class="n">gen_dir</span><span class="p">,</span>
        <span class="n">flank_transposon</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;flank_&#39;</span> <span class="o">+</span> <span class="n">tr_id</span><span class="p">]</span> <span class="o">*</span> <span class="n">strand_transposon</span> <span class="o">*</span> <span class="n">tr_dir</span><span class="p">,</span>
        <span class="n">strand_genome</span><span class="o">=</span><span class="n">strand_genome</span><span class="p">,</span>
        <span class="n">strand_transposon</span><span class="o">=</span><span class="n">strand_transposon</span><span class="p">,</span>
        <span class="n">support_junction</span><span class="o">=</span><span class="n">support_junction</span><span class="p">,</span>
        <span class="n">support_spanning</span><span class="o">=</span><span class="n">support_spanning</span><span class="p">,</span>
        <span class="n">metadata</span><span class="o">=</span><span class="n">frozendict</span><span class="p">())</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Julian de Ruiter.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../../',
            VERSION:'0.3.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>