

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>imfusion.aligners.tophat2.insertions &mdash; IM-Fusion 0.1+1.g9b7cc62 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="IM-Fusion 0.1+1.g9b7cc62 documentation" href="../../../../index.html"/>
        <link rel="up" title="Module code" href="../../../index.html"/> 

  
  <script src="../../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../../index.html" class="icon icon-home"> IM-Fusion
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../home.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../command_line_usage.html">Command-line usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../development.html">Development</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../../index.html">IM-Fusion</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../../index.html">Module code</a> &raquo;</li>
      
    <li>imfusion.aligners.tophat2.insertions</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for imfusion.aligners.tophat2.insertions</h1><div class="highlight"><pre>
<span></span><span class="c1"># pylint: disable=W0622,W0614,W0401</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">absolute_import</span><span class="p">,</span> <span class="n">division</span><span class="p">,</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">builtins</span> <span class="kn">import</span> <span class="o">*</span>
<span class="c1"># pylint: enable=W0622,W0614,W0401</span>

<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="k">try</span><span class="p">:</span>
    <span class="c1"># Python 3.0+</span>
    <span class="kn">from</span> <span class="nn">subprocess</span> <span class="kn">import</span> <span class="n">DEVNULL</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="c1"># Python 2.x</span>
    <span class="n">DEVNULL</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">devnull</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">pathlib2</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">imfusion.model</span> <span class="kn">import</span> <span class="n">Fusion</span>

<span class="kn">from</span> <span class="nn">imfusion.util.check</span> <span class="kn">import</span> <span class="n">check_features</span>
<span class="kn">from</span> <span class="nn">imfusion.util.fusions</span> <span class="kn">import</span> <span class="n">annotate_fusions</span><span class="p">,</span> <span class="n">place_fusions</span>
<span class="kn">from</span> <span class="nn">imfusion.util.insertions</span> <span class="kn">import</span> <span class="n">filter_invalid_insertions</span>
<span class="kn">from</span> <span class="nn">imfusion.util.shell</span> <span class="kn">import</span> <span class="n">format_kwargs</span>
<span class="kn">from</span> <span class="nn">imfusion.util.tabix</span> <span class="kn">import</span> <span class="n">GtfFile</span>


<div class="viewcode-block" id="identify_insertions"><a class="viewcode-back" href="../../../../api.html#imfusion.aligners.tophat2.identify_insertions">[docs]</a><span class="k">def</span> <span class="nf">identify_insertions</span><span class="p">(</span><span class="n">fastqs</span><span class="p">,</span> <span class="n">index_path</span><span class="p">,</span> <span class="n">reference_gtf_path</span><span class="p">,</span>
                        <span class="n">transposon_name</span><span class="p">,</span> <span class="n">transposon_features</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span>
                        <span class="n">work_dir</span><span class="p">,</span> <span class="n">min_flank</span><span class="p">,</span> <span class="n">tophat_kws</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">transcriptome_index</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Identifies insertions from RNA-seq fusions using Tophat2.</span>

<span class="sd">    Main function for identifying fusions from RNA-seq fastq files using Tophat2. The function essentially consists of four main steps:</span>

<span class="sd">        - The identification of gene-transposon fusions using Tophat2</span>
<span class="sd">        - Annotation of the found fusions for gene/transposon features</span>
<span class="sd">        - Deriving approximate locations for the corresponding insertions.</span>
<span class="sd">        - Filtering of fusions that are biologically implausible (for</span>
<span class="sd">          example due to their relative orientation)</span>

<span class="sd">    The function returns the list of insertions that were identified by</span>
<span class="sd">    Tophat2. The generated alignment is also symlinked into the work directory</span>
<span class="sd">    as &#39;alignment.bam&#39; for convenient access.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fastqs : list[pathlib.Path] or list[tuple(pathlib.Path, pathlib.Path)]</span>
<span class="sd">        Paths to the fastq files that should be used for the Tophat2</span>
<span class="sd">        alignment. Can be given as a list of file paths for single-end</span>
<span class="sd">        sequencing data, or a list of path tuples for paired-end sequencing</span>
<span class="sd">        data. The fastqs are treated as belonging to a single sample.</span>
<span class="sd">    index_path : pathlib.Path</span>
<span class="sd">        Path to the bowtie index of the (augmented)</span>
<span class="sd">        genome that should be used in the alignment. This index is</span>
<span class="sd">        typically generated by the *build_reference* function.</span>
<span class="sd">    reference_gtf_path : pathlib.Path:</span>
<span class="sd">        Path to the gtf file containing genomic</span>
<span class="sd">        features. This file is used by Tophat2 for known gene features and</span>
<span class="sd">        for the annotation of gene features for identified fusions.</span>
<span class="sd">    transposon_name : str</span>
<span class="sd">        Name of the transposon sequence in the augmented reference genome.</span>
<span class="sd">    transposon_features :pandas.DataFrame</span>
<span class="sd">        Dataframe containing positions</span>
<span class="sd">        for the features present in the transposon sequence. Used to</span>
<span class="sd">        identify transposon features (such as splice acceptors or donors)</span>
<span class="sd">        that are involed in the identified fusions.</span>
<span class="sd">    sample_id : str</span>
<span class="sd">        Sample name that the identified insertions should be assigned to.</span>
<span class="sd">    work_dir : pathlib.Path</span>
<span class="sd">        Path to the working directory.</span>
<span class="sd">    min_flank : int</span>
<span class="sd">        Minimum amount of flanking region that should be</span>
<span class="sd">        surrounding the fusion. Used by Tophat2 in its identification</span>
<span class="sd">        of fusions during the alignment.</span>
<span class="sd">    tophat_kws : dict</span>
<span class="sd">        Dict of extra arguments for Tophat2.</span>

<span class="sd">    Yields</span>
<span class="sd">    -------</span>
<span class="sd">    Insertion</span>
<span class="sd">            Next insertion that was identified in the given sample.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Sanity checks for inputs.</span>
    <span class="c1"># TODO: check GTF?</span>
    <span class="n">check_features</span><span class="p">(</span><span class="n">transposon_features</span><span class="p">)</span>

    <span class="c1"># Identify fusions.</span>
    <span class="n">fusions</span> <span class="o">=</span> <span class="n">identify_fusions</span><span class="p">(</span>
        <span class="n">fastqs</span><span class="p">,</span>
        <span class="n">index_path</span><span class="o">=</span><span class="n">index_path</span><span class="p">,</span>
        <span class="n">transposon_name</span><span class="o">=</span><span class="n">transposon_name</span><span class="p">,</span>
        <span class="n">reference_gtf_path</span><span class="o">=</span><span class="n">reference_gtf_path</span><span class="p">,</span>
        <span class="n">work_dir</span><span class="o">=</span><span class="n">work_dir</span><span class="p">,</span>
        <span class="n">min_flank</span><span class="o">=</span><span class="n">min_flank</span><span class="p">,</span>
        <span class="n">tophat_kws</span><span class="o">=</span><span class="n">tophat_kws</span><span class="p">,</span>
        <span class="n">transcriptome_index</span><span class="o">=</span><span class="n">transcriptome_index</span><span class="p">)</span>

    <span class="c1"># Annotate fusions for gene/transposon features and strands.</span>
    <span class="n">reference_gtf</span> <span class="o">=</span> <span class="n">GtfFile</span><span class="p">(</span><span class="n">reference_gtf_path</span><span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;-- Annotating gene-transposon fusions&#39;</span><span class="p">)</span>
    <span class="n">annotated_fusions</span> <span class="o">=</span> <span class="n">annotate_fusions</span><span class="p">(</span>
        <span class="n">fusions</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">,</span> <span class="n">transposon_features</span><span class="p">)</span>

    <span class="c1"># Filter insertions without or with an unknown transposon feature.</span>
    <span class="n">filtered_fusions</span> <span class="o">=</span> <span class="p">(</span><span class="n">fusion</span> <span class="k">for</span> <span class="n">fusion</span> <span class="ow">in</span> <span class="n">annotated_fusions</span>
                        <span class="k">if</span> <span class="n">fusion</span><span class="o">.</span><span class="n">feature_type</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;SD&#39;</span><span class="p">,</span> <span class="s1">&#39;SA&#39;</span><span class="p">})</span>

    <span class="c1"># Determine approximate genomic position for insertions.</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;-- Converting to insertions&#39;</span><span class="p">)</span>
    <span class="n">insertions</span> <span class="o">=</span> <span class="n">place_fusions</span><span class="p">(</span><span class="n">filtered_fusions</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">reference_gtf</span><span class="p">)</span>

    <span class="c1"># Filter wrong insertions.</span>
    <span class="n">insertions</span> <span class="o">=</span> <span class="n">filter_invalid_insertions</span><span class="p">(</span><span class="n">insertions</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">insertions</span></div>


<span class="k">def</span> <span class="nf">identify_fusions</span><span class="p">(</span><span class="n">fastqs</span><span class="p">,</span> <span class="n">index_path</span><span class="p">,</span> <span class="n">reference_gtf_path</span><span class="p">,</span> <span class="n">transposon_name</span><span class="p">,</span>
                     <span class="n">work_dir</span><span class="p">,</span> <span class="n">min_flank</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">use_existing</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                     <span class="n">tophat_kws</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">transcriptome_index</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Identifies gene-transposon fusions from RNA-seq data using Tophat2.</span>

<span class="sd">    fastqs : list[pathlib.Path] or list[tuple(pathlib.Path, pathlib.Path)]</span>
<span class="sd">        Paths to the fastq files that should be used for the Tophat2</span>
<span class="sd">        alignment. Can be given as a list of file paths for single-end</span>
<span class="sd">        sequencing data, or a list of path tuples for paired-end sequencing</span>
<span class="sd">        data. The fastqs are treated as belonging to a single sample.</span>
<span class="sd">    index_path : pathlib.Path)</span>
<span class="sd">        Path to the bowtie index of the (augmented)</span>
<span class="sd">        genome that should be used in the alignment. This index is</span>
<span class="sd">        typically generated by the *build_reference* function.</span>
<span class="sd">    reference_gtf_path : pathlib.Path:</span>
<span class="sd">        Path to the gtf file containing genomic</span>
<span class="sd">        features. This file is used by Tophat2 for known gene features and</span>
<span class="sd">        for the annotation of gene features for identified fusions.</span>
<span class="sd">    transposon_name : str</span>
<span class="sd">        Name of the transposon sequence in the augmented reference genome.</span>
<span class="sd">    work_dir : pathlib.Path</span>
<span class="sd">        Path to the working directory.</span>
<span class="sd">    min_flank : int</span>
<span class="sd">        Minimum amount of flanking region that should be</span>
<span class="sd">        surrounding the fusion. Used by Tophat2 in its identification</span>
<span class="sd">        of fusions during the alignment.</span>
<span class="sd">    use_existing : bool</span>
<span class="sd">        Whether to use existing Tophat2 output if present (default = True).</span>
<span class="sd">    tophat_kws : dict</span>
<span class="sd">        Dict of extra arguments for Tophat2.</span>
<span class="sd">    transcriptome_index : pathlib.Path</span>
<span class="sd">        Path to the transcriptome index to use in the alignment. Only needed</span>
<span class="sd">        if different from the default path used by `build_reference`.</span>

<span class="sd">    Yields</span>
<span class="sd">    -------</span>
<span class="sd">    Fusion</span>
<span class="sd">        Next fusion that was identified in the given sample.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">tophat_kws</span> <span class="o">=</span> <span class="n">tophat_kws</span> <span class="ow">or</span> <span class="p">{}</span>

    <span class="k">if</span> <span class="n">transcriptome_index</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="c1"># Check if index exists at default location.</span>
        <span class="n">tr_base</span> <span class="o">=</span> <span class="n">index_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="n">index_path</span><span class="o">.</span><span class="n">suffix</span> <span class="o">+</span> <span class="s1">&#39;.transcriptome&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">tr_base</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="n">tr_base</span><span class="o">.</span><span class="n">suffix</span> <span class="o">+</span> <span class="s1">&#39;.1.ebwt&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">transcriptome_index</span> <span class="o">=</span> <span class="n">tr_base</span>

    <span class="c1"># Use sub-directory as work_dir for tophat.</span>
    <span class="n">tophat_dir</span> <span class="o">=</span> <span class="n">work_dir</span> <span class="o">/</span> <span class="s1">&#39;tophat&#39;</span>

    <span class="c1"># Expected path for fusion file.</span>
    <span class="n">fusion_path</span> <span class="o">=</span> <span class="n">tophat_dir</span> <span class="o">/</span> <span class="s1">&#39;fusions.out&#39;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">use_existing</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">fusion_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;-- Running alignment&#39;</span><span class="p">)</span>
        <span class="c1"># Perform alignment to identify fusions.</span>

        <span class="c1"># Add our arguments to any passed kwargs.</span>
        <span class="n">tophat_kws</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s1">&#39;--fusion-search&#39;</span><span class="p">:</span> <span class="bp">True</span><span class="p">,</span>
                           <span class="s1">&#39;--fusion-anchor-length&#39;</span><span class="p">:</span> <span class="n">min_flank</span><span class="p">,</span>
                           <span class="s1">&#39;--bowtie1&#39;</span><span class="p">:</span> <span class="bp">True</span><span class="p">})</span>

        <span class="c1"># Inject index or gtf if given.</span>
        <span class="k">if</span> <span class="n">transcriptome_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">tophat_kws</span><span class="p">[</span><span class="s1">&#39;--transcriptome-index&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">transcriptome_index</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">reference_gtf_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">tophat_kws</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;-G&#39;</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>
            <span class="n">tophat_kws</span><span class="p">[</span><span class="s1">&#39;--GTF&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">reference_gtf_path</span><span class="p">)</span>

        <span class="c1"># Do alignment with Tophat2.</span>
        <span class="n">tophat2_align</span><span class="p">(</span><span class="n">fastqs</span><span class="o">=</span><span class="n">fastqs</span><span class="p">,</span> <span class="n">output_dir</span><span class="o">=</span><span class="n">tophat_dir</span><span class="p">,</span>
                      <span class="n">index_path</span><span class="o">=</span><span class="n">index_path</span><span class="p">,</span> <span class="n">kwargs</span><span class="o">=</span><span class="n">tophat_kws</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;-- Using existing tophat alignment&#39;</span><span class="p">)</span>

    <span class="c1"># Symlink alignment.</span>
    <span class="n">aln_src_path</span> <span class="o">=</span> <span class="n">tophat_dir</span> <span class="o">/</span> <span class="s1">&#39;accepted_hits.bam&#39;</span>
    <span class="n">aln_tgt_path</span> <span class="o">=</span> <span class="n">work_dir</span> <span class="o">/</span> <span class="s1">&#39;alignment.bam&#39;</span>

    <span class="k">if</span> <span class="n">aln_tgt_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">aln_tgt_path</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>

    <span class="n">aln_tgt_path</span><span class="o">.</span><span class="n">symlink_to</span><span class="p">(</span><span class="n">aln_src_path</span><span class="o">.</span><span class="n">relative_to</span><span class="p">(</span><span class="n">aln_tgt_path</span><span class="o">.</span><span class="n">parent</span><span class="p">))</span>

    <span class="c1"># Extract fusions.</span>
    <span class="n">fusions</span> <span class="o">=</span> <span class="n">extract_fusions</span><span class="p">(</span><span class="n">fusion_path</span><span class="p">,</span> <span class="n">transposon_name</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">fusions</span>


<span class="k">def</span> <span class="nf">_has_index</span><span class="p">(</span><span class="n">base_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns True if index exists.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">Path</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">base_path</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;.1.ebwt&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">extract_fusions</span><span class="p">(</span><span class="n">fusion_path</span><span class="p">,</span> <span class="n">transposon_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extracts gene-transposon fusions from a Tophat fusion.out file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file_path : pathlib.Path</span>
<span class="sd">        Path to the Tophat fusion file (fusions.out).</span>
<span class="sd">    transposon_name : str</span>
<span class="sd">        Name of the transposon sequence in the augmented reference</span>
<span class="sd">        genome that was used for the alignment.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pandas.DataFrame</span>
<span class="sd">        DataFrame containing gene-transposon fusions.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Read fusions and extract insertions.</span>
    <span class="n">fusions</span> <span class="o">=</span> <span class="n">read_fusions</span><span class="p">(</span><span class="n">fusion_path</span><span class="p">)</span>

    <span class="c1"># Select fusions where one seqname is the transposon and the other isn&#39;t.</span>
    <span class="n">fusions</span> <span class="o">=</span> <span class="n">fusions</span><span class="o">.</span><span class="n">ix</span><span class="p">[((</span><span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;seqname_a&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">transposon_name</span><span class="p">)</span> <span class="o">^</span>
                          <span class="p">(</span><span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;seqname_b&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">transposon_name</span><span class="p">))]</span>

    <span class="c1"># Build frame with candidate insertions.</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">fusion_row</span> <span class="ow">in</span> <span class="n">fusions</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">yield</span> <span class="n">_to_fusion_obj</span><span class="p">(</span><span class="n">fusion_row</span><span class="p">,</span> <span class="n">transposon_name</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_to_fusion_obj</span><span class="p">(</span><span class="n">fusion</span><span class="p">,</span> <span class="n">transposon_name</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">fusion</span><span class="o">.</span><span class="n">seqname_a</span> <span class="o">==</span> <span class="n">transposon_name</span><span class="p">:</span>
        <span class="n">gen_id</span><span class="p">,</span> <span class="n">tr_id</span> <span class="o">=</span> <span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gen_id</span><span class="p">,</span> <span class="n">tr_id</span> <span class="o">=</span> <span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span>

    <span class="k">return</span> <span class="n">Fusion</span><span class="p">(</span><span class="n">seqname</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;seqname_&#39;</span> <span class="o">+</span> <span class="n">gen_id</span><span class="p">],</span>
                  <span class="n">anchor_genome</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;location_&#39;</span> <span class="o">+</span> <span class="n">gen_id</span><span class="p">],</span>
                  <span class="n">anchor_transposon</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;location_&#39;</span> <span class="o">+</span> <span class="n">tr_id</span><span class="p">],</span>
                  <span class="n">flank_genome</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;flank_&#39;</span> <span class="o">+</span> <span class="n">gen_id</span><span class="p">],</span>
                  <span class="n">flank_transposon</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;flank_&#39;</span> <span class="o">+</span> <span class="n">tr_id</span><span class="p">],</span>
                  <span class="n">strand_genome</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;strand_&#39;</span> <span class="o">+</span> <span class="n">gen_id</span><span class="p">],</span>
                  <span class="n">strand_transposon</span><span class="o">=</span><span class="n">fusion</span><span class="p">[</span><span class="s1">&#39;strand_&#39;</span> <span class="o">+</span> <span class="n">tr_id</span><span class="p">],</span>
                  <span class="n">spanning_reads</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">supp_reads</span><span class="p">,</span>
                  <span class="n">supporting_mates</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">supp_mates</span><span class="p">,</span>
                  <span class="n">spanning_mates</span><span class="o">=</span><span class="n">fusion</span><span class="o">.</span><span class="n">supp_spanning_mates</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">read_fusions</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads fusion.out file from Tophat2.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file_path : pathlib.Path</span>
<span class="sd">        Path to the Tophat fusion file (fusions.out).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pandas.DataFrame</span>
<span class="sd">        DataFrame containing gene fusions.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Read fusions using pandas.</span>
    <span class="n">col_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;seqnames&#39;</span><span class="p">,</span> <span class="s1">&#39;location_a&#39;</span><span class="p">,</span> <span class="s1">&#39;location_b&#39;</span><span class="p">,</span> <span class="s1">&#39;orientation&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;supp_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;supp_mates&#39;</span><span class="p">,</span> <span class="s1">&#39;supp_spanning_mates&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;contradicting_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;flank_a&#39;</span><span class="p">,</span> <span class="s1">&#39;flank_b&#39;</span><span class="p">]</span>

    <span class="n">fusions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">file_path</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                          <span class="n">usecols</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">names</span><span class="o">=</span><span class="n">col_names</span><span class="p">)</span>

    <span class="c1"># Split combined entries.</span>
    <span class="n">fusions</span> <span class="o">=</span> <span class="n">_split_fields</span><span class="p">(</span><span class="n">fusions</span><span class="p">)</span>

    <span class="c1"># Map orientation to strands.</span>
    <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_a&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">({</span><span class="s1">&#39;f&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">})</span>
    <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_b&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_b&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">({</span><span class="s1">&#39;f&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">})</span>

    <span class="k">return</span> <span class="n">fusions</span>


<span class="k">def</span> <span class="nf">_split_fields</span><span class="p">(</span><span class="n">fusions</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Splits combined seqnames/strand entries in tophat fusion frame.&quot;&quot;&quot;</span>

    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;seqname_a&#39;</span><span class="p">,</span> <span class="s1">&#39;location_a&#39;</span><span class="p">,</span> <span class="s1">&#39;strand_a&#39;</span><span class="p">,</span>
               <span class="s1">&#39;seqname_b&#39;</span><span class="p">,</span> <span class="s1">&#39;location_b&#39;</span><span class="p">,</span> <span class="s1">&#39;strand_b&#39;</span><span class="p">,</span>
               <span class="s1">&#39;supp_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;supp_mates&#39;</span><span class="p">,</span> <span class="s1">&#39;supp_spanning_mates&#39;</span><span class="p">,</span>
               <span class="s1">&#39;contradicting_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;flank_a&#39;</span><span class="p">,</span> <span class="s1">&#39;flank_b&#39;</span><span class="p">]</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fusions</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Return empty &#39;split&#39; frame.</span>
        <span class="n">fusions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">([],</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Split combined entries.</span>
        <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;seqname_a&#39;</span><span class="p">],</span> <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;seqname_b&#39;</span><span class="p">]</span> <span class="o">=</span> \
            <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;seqnames&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">))</span>

        <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_a&#39;</span><span class="p">],</span> <span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;strand_b&#39;</span><span class="p">]</span> <span class="o">=</span> \
            <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">fusions</span><span class="p">[</span><span class="s1">&#39;orientation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">list</span><span class="p">))</span>

        <span class="c1"># Subset/Reorder columns.</span>
        <span class="n">fusions</span> <span class="o">=</span> <span class="n">fusions</span><span class="p">[</span><span class="n">columns</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">fusions</span>


<span class="k">def</span> <span class="nf">tophat2_align</span><span class="p">(</span><span class="n">fastqs</span><span class="p">,</span> <span class="n">index_path</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">kwargs</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                  <span class="n">path</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">log_path</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">check_python</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Aligns fastq files to a reference genome using TopHat2.</span>

<span class="sd">    This function is used to call TopHat2 from Python to perform an</span>
<span class="sd">    RNA-seq alignment. As Tophat2 is written in Python 2.7, this function</span>
<span class="sd">    cannot be used in Python 3.0+.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fastqs : list[pathlib.Path] or list[tuple(pathlib.Path, pathlib.Path)]</span>
<span class="sd">        Paths to the fastq files that should be used for the Tophat2</span>
<span class="sd">        alignment. Can be given as a list of file paths for single-end</span>
<span class="sd">        sequencing data, or a list of path tuples for paired-end sequencing</span>
<span class="sd">        data. The fastqs are treated as belonging to a single sample.</span>
<span class="sd">    index_path : pathlib.Path)</span>
<span class="sd">        Path to the bowtie index of the (augmented)</span>
<span class="sd">        genome that should be used in the alignment. This index is</span>
<span class="sd">        typically generated by the *build_reference* function.</span>
<span class="sd">    output_dir : pathlib.Path</span>
<span class="sd">            Path to the output directory.</span>
<span class="sd">    kwargs : dict</span>
<span class="sd">        Dict of extra command line arguments for Tophat2.</span>
<span class="sd">    path : pathlib.Path</span>
<span class="sd">        Path to the Tophat2 executable.</span>
<span class="sd">    check_python (bool): Whether to check if we are running on Python 2.</span>
<span class="sd">        Raises ValueError if this is not the case and check_python is True.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Path to output directory (containing the alignment).</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">tophat_kwargs</span> <span class="o">=</span> <span class="n">kwargs</span> <span class="ow">or</span> <span class="nb">dict</span><span class="p">()</span>

    <span class="c1"># Check Python version.</span>
    <span class="k">if</span> <span class="n">check_python</span> <span class="ow">and</span> <span class="n">sys</span><span class="o">.</span><span class="n">version_info</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Python 3.x is not supported for identifying &#39;</span>
                         <span class="s1">&#39;insertions as TopHat2 does not support Python 3&#39;</span><span class="p">)</span>

    <span class="c1"># Create output_dir if needed.</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">output_dir</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">output_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># Extract fastq paths and concatenate into single str.</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fastqs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">tuple</span><span class="p">):</span>
        <span class="n">fastqs_1</span><span class="p">,</span> <span class="n">fastqs_2</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">fastqs</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">fastqs_1</span><span class="p">,</span> <span class="n">fastqs_2</span> <span class="o">=</span> <span class="n">fastqs</span><span class="p">,</span> <span class="bp">None</span>

    <span class="c1"># Concatenate inputs.</span>
    <span class="n">fastqs_1</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">fp</span><span class="p">)</span> <span class="k">for</span> <span class="n">fp</span> <span class="ow">in</span>  <span class="n">fastqs_1</span><span class="p">)</span>
    <span class="n">fastqs_2</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="nb">str</span><span class="p">(</span><span class="n">fp</span><span class="p">)</span> <span class="k">for</span> <span class="n">fp</span> <span class="ow">in</span> <span class="n">fastqs_2</span><span class="p">))</span> \
        <span class="k">if</span> <span class="n">fastqs_2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="k">else</span> <span class="bp">None</span>
    
    <span class="c1"># Remove output_dir kwargs, as we override these with our</span>
    <span class="c1"># own arguments anyway. After removing, inject out output_dir.</span>
    <span class="n">overridden_kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="s1">&#39;--output-dir&#39;</span><span class="p">}</span>
    <span class="n">tophat_kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">tophat_kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                     <span class="k">if</span> <span class="n">k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">overridden_kwargs</span><span class="p">)}</span>

    <span class="n">tophat_kwargs</span><span class="p">[</span><span class="s1">&#39;--output-dir&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="c1"># Build command-line arguments.</span>
    <span class="n">tophat2_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">path</span> <span class="ow">or</span> <span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="o">/</span> <span class="s1">&#39;tophat2&#39;</span>
    <span class="n">optional_args</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">format_kwargs</span><span class="p">(</span><span class="n">tophat_kwargs</span><span class="p">))</span>

    <span class="n">positional_args</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">index_path</span><span class="p">),</span> <span class="n">fastqs_1</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">fastqs_2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">positional_args</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fastqs_2</span><span class="p">)</span>

    <span class="n">cmdline_args</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">tophat2_path</span><span class="p">)]</span> <span class="o">+</span> <span class="n">optional_args</span> <span class="o">+</span> <span class="n">positional_args</span>

    <span class="c1"># Run Tophat2!</span>
    <span class="k">if</span> <span class="n">log_path</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">cmdline_args</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">DEVNULL</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">log_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log_file</span><span class="p">:</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">cmdline_args</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">log_file</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">output_dir</span>
</pre></div>

           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Julian de Ruiter.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../../',
            VERSION:'0.1+1.g9b7cc62',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>